%! BibTeX Compiler = biber
%DIF LATEXDIFF DIFFERENCE FILE
%DIF DEL main-revision.tex        Sat Feb  4 22:51:29 2023
%DIF ADD main-revision-rev2.tex   Mon Feb  6 15:36:38 2023
%TC:ignore
\documentclass{article}

\usepackage{xcolor, colortbl}
\definecolor{BLUELINK}{HTML}{0645AD}
\definecolor{DARKBLUELINK}{HTML}{0B0080}
\definecolor{LIGHTBLUELINK}{HTML}{3366BB}
\definecolor{PURPLELINK}{HTML}{663366}
\definecolor{RED}{HTML}{EB6231}
\definecolor{GREEN}{HTML}{8FB03E}
\definecolor{LIGHTGREY}{gray}{0.9}
\PassOptionsToPackage{hyphens}{url}
\usepackage[colorlinks=false]{hyperref}
% for linking between references, figures, TOC, etc in the pdf document
\hypersetup{colorlinks,
    linkcolor=DARKBLUELINK,
    anchorcolor=DARKBLUELINK,
    citecolor=DARKBLUELINK,
    filecolor=DARKBLUELINK,
    menucolor=DARKBLUELINK,
    urlcolor=BLUELINK
} % Color citation links in purple
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{naturalnames}{hyperref}

\usepackage{biorxiv}
\usepackage[backend=biber,eprint=false,isbn=false,url=false,intitle=true,style=nature,date=year]{biblatex}
\addbibresource{codon_models.bib}

\usepackage{url}
\usepackage{amssymb,amsfonts,amsmath,amsthm,mathtools}
\usepackage{textcomp}
\usepackage{gensymb}
\usepackage{cancel}
\usepackage{lmodern}
\usepackage{xfrac, nicefrac}
\usepackage{blkarray}
\usepackage{pgf,tikz}
\usetikzlibrary{positioning,arrows,automata,calc}
\usepackage{bm}
\usepackage{icomma}
\usepackage{listings, enumerate, enumitem}
\usepackage[export]{adjustbox}
\usepackage{graphicx}
\usepackage{tabu}
\usepackage{hhline}
\usepackage{multicol,multirow,array}
\usepackage{etoolbox}
\usepackage{booktabs}

\usepackage[flushleft] {threeparttable}
\usepackage{bbold}
\usepackage{pdfpages}
\usetikzlibrary{positioning}
\usetikzlibrary{arrows,automata}
\pdfinclusioncopyfonts=1
\usepackage{nicefrac} % compact symbols for 1/2, etc.
\usepackage{microtype} % microtypography
\usepackage{lineno}

\graphicspath{{artworks/}}
\makeatletter
\def\input@path{{artworks/}}
\makeatother

\newcommand{\specialcell}[2][c]{%
    \begin{tabular}[#1]{@{}c@{}}
        #2
    \end{tabular}}

\newcommand{\UniDimArray}[1]{\bm{#1}}
\newcommand{\BiDimArray}[1]{\bm{#1}}

\DeclareMathOperator{\E}{\mathbb{E}}
\DeclareMathOperator{\Var}{\mathrm{Var}}
\newcommand{\der}{\mathrm{d}}
\newcommand{\angstrom}{\mathrm{\normalfont\AA}}
\newcommand{\e}{\mathrm{e}}
\newcommand{\avg}[1]{\left< #1 \right>} % for average
\newcommand{\Ne}{N_{\mathrm{e}}}
\newcommand{\dn}{d_N}
\newcommand{\ds}{d_S}
\newcommand{\dnds}{\dn / \ds}
\newcommand{\rateApop}{\omega_{\mathrm{A}}}
\newcommand{\rateAphy}{\rateApop^{\mathrm{phy}}}
\newcommand{\pn}{\pi_N}
\newcommand{\ps}{\pi_S}
\newcommand{\pnps}{\pn / \ps}
\newcommand{\proba}{\mathbb{P}}
\newcommand{\pfix}{\proba_{\mathrm{fix}}}
\newcommand{\Pfix}{2 \Ne \proba_{\mathrm{fix}}}
\newcommand{\indice}{a}
\newcommand{\indiceexp}{^{(\indice)}}

\renewcommand{\baselinestretch}{1.25}
\renewcommand{\arraystretch}{0.6}
\linenumbers

\title{Genes and sites under adaptation at the phylogenetic scale also exhibit adaptation at the population-genetic scale}

%\date{September 9, 1985}	% Here you can change the date presented in the paper title
%\date{} \rowcolor{LIGHTGREY}% Or removing it

\author{
    \large
    \textbf{T. {Latrille}$^{1,2,3}$, N. {Rodrigue}$^{4}$, N. {Lartillot}$^{1}$}\\
    \normalsize
    $^{1}$Université de Lyon, Université Lyon 1, CNRS, VetAgro Sup, Laboratoire de Biométrie et Biologie Evolutive, UMR5558, Villeurbanne, France\\
    $^{2}$École Normale Supérieure de Lyon, Université de Lyon, Lyon, France\\
    $^{3}$Department of Computational Biology, Université de Lausanne, Lausanne, Switzerland\\
    $^{4}$Department of Biology, Institute of Biochemistry, and School of Mathematics and Statistics, Carleton University, Ottawa, Canada \\
    \texttt{\href{mailto:thibault.latrille@ens-lyon.org}{thibault.latrille@ens-lyon.org}} \\
}

%TC:endignore
%DIF PREAMBLE EXTENSION ADDED BY LATEXDIFF
%DIF UNDERLINE PREAMBLE %DIF PREAMBLE
\RequirePackage[normalem]{ulem} %DIF PREAMBLE
\RequirePackage{color}\definecolor{RED}{rgb}{1,0,0}\definecolor{BLUE}{rgb}{0,0,1} %DIF PREAMBLE
\providecommand{\DIFaddtex}[1]{{\protect\color{blue}\uwave{#1}}} %DIF PREAMBLE
\providecommand{\DIFdeltex}[1]{{\protect\color{red}\sout{#1}}}                      %DIF PREAMBLE
%DIF SAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddbegin}{} %DIF PREAMBLE
\providecommand{\DIFaddend}{} %DIF PREAMBLE
\providecommand{\DIFdelbegin}{} %DIF PREAMBLE
\providecommand{\DIFdelend}{} %DIF PREAMBLE
\providecommand{\DIFmodbegin}{} %DIF PREAMBLE
\providecommand{\DIFmodend}{} %DIF PREAMBLE
%DIF FLOATSAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddFL}[1]{\DIFadd{#1}} %DIF PREAMBLE
\providecommand{\DIFdelFL}[1]{\DIFdel{#1}} %DIF PREAMBLE
\providecommand{\DIFaddbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFaddendFL}{} %DIF PREAMBLE
\providecommand{\DIFdelbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFdelendFL}{} %DIF PREAMBLE
%DIF HYPERREF PREAMBLE %DIF PREAMBLE
\providecommand{\DIFadd}[1]{\texorpdfstring{\DIFaddtex{#1}}{#1}} %DIF PREAMBLE
\providecommand{\DIFdel}[1]{\texorpdfstring{\DIFdeltex{#1}}{}} %DIF PREAMBLE
\newcommand{\DIFscaledelfig}{0.5}
%DIF HIGHLIGHTGRAPHICS PREAMBLE %DIF PREAMBLE
\RequirePackage{settobox} %DIF PREAMBLE
\RequirePackage{letltxmacro} %DIF PREAMBLE
\newsavebox{\DIFdelgraphicsbox} %DIF PREAMBLE
\newlength{\DIFdelgraphicswidth} %DIF PREAMBLE
\newlength{\DIFdelgraphicsheight} %DIF PREAMBLE
% store original definition of \includegraphics %DIF PREAMBLE
\LetLtxMacro{\DIFOincludegraphics}{\includegraphics} %DIF PREAMBLE
\newcommand{\DIFaddincludegraphics}[2][]{{\color{blue}\fbox{\DIFOincludegraphics[#1]{#2}}}} %DIF PREAMBLE
\newcommand{\DIFdelincludegraphics}[2][]{% %DIF PREAMBLE
\sbox{\DIFdelgraphicsbox}{\DIFOincludegraphics[#1]{#2}}% %DIF PREAMBLE
\settoboxwidth{\DIFdelgraphicswidth}{\DIFdelgraphicsbox} %DIF PREAMBLE
\settoboxtotalheight{\DIFdelgraphicsheight}{\DIFdelgraphicsbox} %DIF PREAMBLE
\scalebox{\DIFscaledelfig}{% %DIF PREAMBLE
\parbox[b]{\DIFdelgraphicswidth}{\usebox{\DIFdelgraphicsbox}\\[-\baselineskip] \rule{\DIFdelgraphicswidth}{0em}}\llap{\resizebox{\DIFdelgraphicswidth}{\DIFdelgraphicsheight}{% %DIF PREAMBLE
\setlength{\unitlength}{\DIFdelgraphicswidth}% %DIF PREAMBLE
\begin{picture}(1,1)% %DIF PREAMBLE
\thicklines\linethickness{2pt} %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,0){\framebox(1,1){}}}% %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,0){\line( 1,1){1}}}% %DIF PREAMBLE
{\color[rgb]{1,0,0}\put(0,1){\line(1,-1){1}}}% %DIF PREAMBLE
\end{picture}% %DIF PREAMBLE
}\hspace*{3pt}}} %DIF PREAMBLE
} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddbegin}{\DIFaddbegin} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddend}{\DIFaddend} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelbegin}{\DIFdelbegin} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelend}{\DIFdelend} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddbegin}{\DIFOaddbegin \let\includegraphics\DIFaddincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddend}{\DIFOaddend \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelbegin}{\DIFOdelbegin \let\includegraphics\DIFdelincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelend}{\DIFOaddend \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddbeginFL}{\DIFaddbeginFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOaddendFL}{\DIFaddendFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelbeginFL}{\DIFdelbeginFL} %DIF PREAMBLE
\LetLtxMacro{\DIFOdelendFL}{\DIFdelendFL} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddbeginFL}{\DIFOaddbeginFL \let\includegraphics\DIFaddincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFaddendFL}{\DIFOaddendFL \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelbeginFL}{\DIFOdelbeginFL \let\includegraphics\DIFdelincludegraphics} %DIF PREAMBLE
\DeclareRobustCommand{\DIFdelendFL}{\DIFOaddendFL \let\includegraphics\DIFOincludegraphics} %DIF PREAMBLE
%DIF LISTINGS PREAMBLE %DIF PREAMBLE
\RequirePackage{listings} %DIF PREAMBLE
\RequirePackage{color} %DIF PREAMBLE
\lstdefinelanguage{DIFcode}{ %DIF PREAMBLE
%DIF DIFCODE_UNDERLINE %DIF PREAMBLE
  moredelim=[il][\color{red}\sout]{\%DIF\ <\ }, %DIF PREAMBLE
  moredelim=[il][\color{blue}\uwave]{\%DIF\ >\ } %DIF PREAMBLE
} %DIF PREAMBLE
\lstdefinestyle{DIFverbatimstyle}{ %DIF PREAMBLE
	language=DIFcode, %DIF PREAMBLE
	basicstyle=\ttfamily, %DIF PREAMBLE
	columns=fullflexible, %DIF PREAMBLE
	keepspaces=true %DIF PREAMBLE
} %DIF PREAMBLE
\lstnewenvironment{DIFverbatim}{\lstset{style=DIFverbatimstyle}}{} %DIF PREAMBLE
\lstnewenvironment{DIFverbatim*}{\lstset{style=DIFverbatimstyle,showspaces=true}}{} %DIF PREAMBLE
%DIF END PREAMBLE EXTENSION ADDED BY LATEXDIFF

\begin{document}
    \maketitle

    \begin{abstract}
        Adaptation in protein-coding sequences can be detected from multiple sequence alignments across species, or alternatively by leveraging polymorphism data within a population.
        Across species, quantification of the adaptive rate relies on phylogenetic codon models, classically formulated in terms of the ratio of non-synonymous over synonymous substitution rates.
        Evidence of an accelerated non-synonymous substitution rate is considered a signature of pervasive adaptation.
        However, because of the background of purifying selection, these models are potentially limited in their sensitivity.
        Recent developments have led to more sophisticated mutation-selection codon models aimed at making a more detailed quantitative assessment of the interplay between mutation, purifying and positive selection.
        In this study, we conducted a large-scale exome-wide analysis of placental mammals with mutation-selection models, assessing their performance at detecting proteins and sites under adaptation.
        Importantly, mutation-selection codon models are based on a population-genetic formalism and thus are directly comparable to McDonald \& Kreitman \DIFdelbegin \DIFdel{tests }\DIFdelend \DIFaddbegin \DIFadd{test }\DIFaddend at the population level to quantify adaptation.
        Taking advantage of this relationship between phylogenetic and population genetics \DIFaddbegin \DIFadd{analyses}\DIFaddend , we integrated divergence and polymorphism data across the entire exome for 29 populations across 7 genera, and showed that proteins and sites detected to be under adaptation at the phylogenetic scale are also under adaptation at the population-genetic scale.
        Altogether, our exome-wide analysis shows that phylogenetic mutation-selection codon models and population-genetic test of adaptation can be reconciled and are congruent, paving the way for integrative models and analyses across individuals and populations.
    \end{abstract}

    \keywords{Adaptation \and phylogenetic \and population genetics \and codon models}

    \section*{Significance Statement}
    Detecting genes under adaptation represents a key step in the decoding of genomes.
    Several methods have been proposed, focussing either on the short time scale (population genetics, e.g. human populations), or on the long time scale (phylogenetics, e.g. across mammals).
    However, the accuracy of these methods is still under debate, and it is still unclear whether the signatures of adaptation are congruent across evolutionary scales.
    In this study, using novel phylogenetic methods and gathering genome data across and within species, we show that the signatures of adaptation at the phylogenetic and population-genetic scales can be reconciled.
    While providing a mutual confirmation of the two approaches, our work paves the way for further methodological integration between micro- and macro-evolutionary genomics.

    \section*{Introduction}
    Present-day genetic sequences are informative of populations' past evolutionary history and can carry signatures of selection at different scales.
    One main goal in the study of molecular evolution is to disentangle and quantify the intensity of neutral, adaptive and purifying evolution acting on sequences, leveraging variations in sequences between and within species.
    Theoretically, in order to detect adaptive evolution, one must have data where part of the sequence is known to be under a neutral regime, which can be used as a null model.
    In the case of protein-coding DNA sequences, synonymous sites are usually taken as proxies for neutral sites, although there are instance where they are indeed under selection\cite{duret_expression_1999, duret_evolution_2002, galtier_codon_2018}.
    Non-synonymous mutations, on the other hand, might be under a mixture of varying degrees of adaptive and purifying selection.
    Contrasting synonymous and non-synonymous changes, two different types of methods have emerged to quantify both positive and purifying selection acting on protein-coding sequences.
    One method, stemming from phylogeny, uses a multiple sequence alignment comprised of genes from different species and relies on codon models to deduce the selective regime from the patterns of non-synonymous versus synonymous substitutions\cite{muse_likelihood_1994,goldman_codonbased_1994}.
    Starting with the work of McDonald \& Kreitman\cite{mcdonald_adaptative_1991}, another method, stemming from population genetics, contrasts polymorphism within a population and divergence to a closely related species.

    At the population-genetic scale, one of the most widely used tests for adaptation relies on the substitutions between two closely related species and polymorphism within one population\cite{mcdonald_adaptative_1991}.
    Under a strict neutral model (i.e.~assuming non-synonymous mutations are either neutral or strongly selected), the ratios of non-synonymous polymorphisms over synonymous polymorphisms ($\pnps$) and of non-synonymous substitutions over synonymous substitutions ($\dnds$) are expected to be equal.
    If, in addition, strongly advantageous mutations occur, they are fixed rapidly in the population, thus contributing solely to divergence but not to polymorphism.
    As a result, the positive difference between $\dnds$ and $\pnps$ gives an estimate of the adaptive rate $\rateApop = \dnds-\pnps$\cite{smith_adaptive_2002}.
    This simple argument is not strictly valid in the presence of moderately deleterious non-synonymous mutations, which can segregate at substantial frequency in the population without reaching fixation, thus contributing solely to polymorphism, and not to divergence, potentially resulting on an underestimation of the rate of adaptive evolution\cite{mcdonald_adaptative_1991, eyre-walker_quantifying_2002}.
    Subsequent developments have tried to correct for this effect by relying on an explicit nearly-neutral model\cite{eyre-walker_estimating_2009, galtier_adaptive_2016}, so as to estimate the rate of evolution expected in the absence of adaptation (called $\omega_0$) based on polymorphism, and then to compare it with the rate of evolution, $\omega=\dnds$, to get an estimate of the rate of adaptation as $\rateApop = \omega-\omega_0$.

    In their current formulation, phylogeny-based methods rely on the ratio of non-synonymous substitutions over synonymous substitutions, called $\omega$\cite{muse_likelihood_1994,goldman_codonbased_1994}.
    Since this ratio is a contrast between the rates before and after the action of selection on the protein, it thus provides an estimate of the strength of selection exerted at the amino-acid level, whether synonymous changes are driven solely by mutation or by both mutation and selection\cite{yang_mutationselection_2008}.
    In this context, an excess in non-synonymous substitutions, leading to $\omega>1$, indicates that the protein is undergoing recurrent positive selection for amino-acids changes, meaning that the protein is putatively under adaptive evolution.
    Conversely, a deficit in non-synonymous substitutions, leading to $\omega<1$, means the protein is under purifying selection.
    In practice, proteins are typically under a mix of adaptive and purifying selection dominated by the latter, thus typically leading to an $\omega<1$ even in the presence of positive selection.
    At a finer scale, site models can detect a specific site ($i$) of the sequence with a $\omega^{(i)}>1$\cite{yang_codonsubstitution_2000, kosiol_patterns_2008}.
    Site models have the advantage of greater sensitivity and the ability to pinpoint where positive selection acts on the protein.
    However, even at the level of a single site under recurrent adaptation, not all \DIFdelbegin \DIFdel{amino-acids }\DIFdelend \DIFaddbegin \DIFadd{amino acids }\DIFaddend are expected to be adaptive, leading to $\omega^{(i)}$ capturing a mix of adaptive and purifying selection, reducing the sensitivity of test.
    An alternative approach to detect adaptation would be to rely on an explicit nearly-neutral model as the null against which to detect deviations, similarly to the McDonald \& Kreitman test.
    Recent development in this direction, the so-called phylogenetic mutation-selection models provide a null model by estimating the fitness landscape over \DIFdelbegin \DIFdel{amino acid }\DIFdelend \DIFaddbegin \DIFadd{amino-acid }\DIFaddend sequences, for each site of the sequence\cite{yang_mutationselection_2008, halpern_evolutionary_1998, rodrigue_mechanistic_2010}.
    At the mutation-selection balance, the probability for a specific codon to be fixed in the population is proportional to its fitness, and a mutation from a high fitness amino acid towards a low fitness amino acid will have a small probability of fixation, genuinely accounting for purifying selection.
    Conversely, only nearly-neutral mutations between high fitness amino acids will tend to be permitted by the model, allowing for the explicit calculation of the nearly-neutral rate of non-synonymous substitutions at mutation-selection balance, called $\omega_{0}$\cite{spielman_relationship_2015, rodrigue_detecting_2017}.
    By contrasting $\omega$ estimated by $\omega$-based codon models and $\omega_{0}$ calculated from mutation-selection models, one can hope to extract the rate of adaptation $\rateAphy = \omega - \omega_{0}$.

    Interestingly, the rate of adaptation is directly comparable between phylogenetic and population-genetic methods since both seek a deviation of $\omega$  from a nearly-neutral null model, estimated with mutation-selection models in phylogenetic context ($\omega_{0}$) or from standing polymorphism in a population-genetic context ($\pnps$).
    This raises the question whether the two signals of adaptation are correlated, thus representing a unique opportunity to confront phylogeny-based and population-based methods.
    These two methods work over very different time scales, for that reason, they might be capturing different signals: long-term evolutionary Red-Queen for phylogeny-based methods versus events of adaptation in specific lineages for population-based methods.
    Nonetheless, we expect sites and proteins under long-term evolutionary Red-Queen regimes to maintain their signal of adaptation in several independent lineages for which the McDonald \& Kreitman test is applied.

    Accordingly, in this study, we first applied $\omega$-based and mutation-selection codon models to whole exome data from placental mammals, so as to quantify the rate $\rateAphy$ for each site and protein and detect signatures of adaptive evolution at the phylogenetic scale.
    Then, we developed a pipeline integrating (and aligning) divergence and polymorphism data across the entire exome for 29 populations across 7 genera, namely \textit{Equus}, \textit{Canis}, \textit{Bos}, \textit{Capra}, \textit{Ovis}, \textit{Chlorocebus} and \textit{Homo}.
    Finally, using this pipeline, we assessed the congruence between the phylogeny-based and population-based approaches, by testing if the group of sequences detected with a high rate of adaptation in the phylogeny-based method also displays a high rate of adaptation according to the population-based method.

    \section*{Results}\label{sec:results}
    \subsection*{Detecting genes and sites under adaptation}
    We derived a two-step approach (see methods), which we applied to a set of alignments of orthologous genes at the scale of placental mammals.
    The $\dnds$ estimated by the site model ($\omega$) is plotted against the $\dnds$ predicted by the nearly-neutral mutation-selection model ($\omega_{0}$) for genes (scatter plot in fig.~\ref{fig:scatterplot}A) and sites (density plot in fig.~\ref{fig:scatterplot}B).
    An excess of $\omega$ relative to $\omega_{0}$ is a typical signature of ongoing positive selection\cite{bloom_identification_2017, rodrigue_detecting_2017}.
    For such comparison to be valid, both models estimating $\omega$ and $\omega_{0}$ should have assumptions as similar as possible regarding the underlying mutation process, modeled as a Muse \& Gaut\cite{muse_likelihood_1994, spielman_relationship_2015}.
    As a control, we validated our Bayesian estimates of $\omega$ against their maximum-likelihood counterpart (methods and fig.~S1).

    Given these validations, an excess of $\omega$ compared to $\omega_{0}$ can be retrieved as a ratio ($\omega / \omega_{0} > 1$) or as a difference ($\omega - \omega_{0}>0$).
    However, since $\omega_{0}$ can be close to $0$ for genes \DIFdelbegin \DIFdel{ans }\DIFdelend \DIFaddbegin \DIFadd{and }\DIFaddend sites strongly constrained (fig.~\ref{fig:scatterplot}), ratios can be particularly large and mathematically undefined, ultimately leading to artifacts.
    Moreover, comparing the posterior credibility intervals of $\omega$ and $\omega_{0}$ as a difference rather than a ratio is more intuitive and visual (fig.~\ref{fig:scatterplot}A).
    Altogether, genes or sites were considered to be under an adaptive regime (in red) if the value of their $\omega$ is higher than that of their $\omega_{0}$, with non-overlapping $95\%$ posterior credibility intervals.
    This procedure retrieved $822$ out \DIFaddbegin \DIFadd{of }\DIFaddend $14,509$ genes, which are putatively under a long-term evolutionary Red-Queen regime.
    At the site level, the nearly-neutral assumption appears to be rejected for $104,129$ out of $8,895,374$ sites.
    Of note, this selection procedure is not meant as a routine statistical test, but only as an enrichment procedure, for the needs of the subsequent analysis shown below.
    In practice, this selection is likely to be conservative and to have a rate of false discovery of the order of 1\% at the gene-level, and 5\% at the site-level (see methods).

    Remarkably, selection based on $\omega>\omega_{0}$ is more sensitive than based on the commonly used criterion of $\omega>1$, since $\omega_{0}$ is always lower than $1$ by definition\cite{spielman_relationship_2015}.
    Thus, we can uncover sites under adaptation ($\omega>\omega_{0}$) with a mean $\omega$ lower than $1$ ($29,543$ sites in fig.~\ref{fig:scatterplot}C).
    These sites could not have been detected by $\omega$-based codon models relying on the criterion that $\omega>1$.
    At the gene level, only two genes have an estimated $\omega > 1$, such that this distinction is not relevant.

    \subsection*{Ontology enrichment tests}
    Next, we investigated whether the genes classified as adaptive ($\omega > \omega_{0}$) showed enrichment in specific ontology terms.
    Thus, we performed $775$ instances of Fisher's exact test to estimate ontology enrichment by contrasting with genes in the control group, not classified as adaptive.
    $42$ ontologies are observed with a p-value ($p_{\mathrm{v}}$) corrected for multiple comparison (Holm–Bonferroni correction, $p_{\mathrm{v}}^{\mathrm{adj}}$) lower than the risk $\alpha=0.05$ (table~S1).
    At a finer scale, we weighted genes by their proportion of sites considered under adaptation with a $\omega$-based site model ($\omega > 1$, table~S2) or with a mutation-selection model ($\omega > \omega_{0}$, table~S3).
    For each ontology, the proportion of sites under adaptation is compared between the set of genes sharing this given ontology and the rest of the genes (Mann-Whitney U test).
    The statistical test based on the first criterion ($\omega>1$) is correlated with ontologies related to immune processes, while the statistical test based on the \DIFdelbegin \DIFdel{the }\DIFdelend second criterion ($\omega > \omega_{0}$) is also correlated with ontologies related to the external membrane and cellular adhesion.

    \subsection*{Congruence between phylogeny- and population-based methods}
    Finally, we investigated whether the phylogeny-based and the population-based methods give congruent results in terms of detection of adaptive evolution (fig.~\ref{fig:method}).
    To do so, population genomic data were collected for 29 populations across 7 genera (fig.~S2).
    For each population, $\rateApop$ \DIFdelbegin \DIFdel{as proposed by }\DIFdelend \DIFaddbegin \DIFadd{based on the }\DIFaddend McDonald \& Kreitman (MK)\cite{mcdonald_adaptative_1991} was computed on the \DIFdelbegin \DIFdel{concatenate }\DIFdelend \DIFaddbegin \DIFadd{concatenated sequence }\DIFaddend of the 822 genes classified as adaptive by the phylogeny-based method (red dots in fig.~\ref{fig:method} and~\ref{fig:unfolded-MK}).
    This result was compared to a null distribution obtained by computing $\rateApop$ over sets of 822 genes that were randomly sampled ($1,000$ replicates) among the genes classifed as nearly-neutral according to the mutation-selection model (green violins in fig.~\ref{fig:method} and~\ref{fig:unfolded-MK}).
    Importantly, the terminal lineages over which the population-genetic method was applied were not included in the phylogenetic analysis, \DIFdelbegin \DIFdel{such }\DIFdelend \DIFaddbegin \DIFadd{so }\DIFaddend as to avoid fallacy and circularity in the estimation of $\rateApop$ and $\rateAphy$.
    As a result, the two methods are working on entirely non-overlapping compartments of the evolutionary history across mammals.
    For all 29 populations, the $\rateApop$ estimated by the population-genetic method was significantly higher for the putatively adaptive gene-set than for the putatively nearly-neutral gene sets of the same size (at a risk $\alpha=0.05$ corrected for multiple testing, Holm-Bonferroni correction).
    There is thus a good qualitative agreement between the two methods as to what they capture and interpret as positive selection at the gene level.

    The same procedure was applied at a finer scale with sites instead of genes.
    For each population, $\rateApop$ was computed on the \DIFdelbegin \DIFdel{concatenate }\DIFdelend \DIFaddbegin \DIFadd{concatenated sequence }\DIFaddend of the $104,129$ sites classified as adaptive by the phylogeny-based method, and compared to the empirical null distribution (fig.~\ref{fig:unfolded-MK}B) and table~\ref{table:unfolded-MK}.
    Out of $29$ populations, $24$ have an $\rateApop$ estimated by the population-genetic method significantly higher for the putatively adaptive site-set than for the putatively nearly-neutral site-sets of the same size taken at random (at a risk $\alpha=0.05$ corrected for multiple testing, Holm-Bonferroni correction).
    In particular, the five populations for which the test is not significant are the human populations.
    Coincidentally, this is the species with the lowest genetic diversity, resulting in under-powered statistical tests due to less SNPs available per loci.
    This statistical effect is combined with the biological effect of increased genetic drift, owing to the small effective population size in ancestral human populations, resulting in less efficient selection.
    Altogether, we argue that finding evidence for positive selection in human populations is statistically more challenging due to both less efficient selection and less data available.

    Except for \textit{Equus} and \textit{Humans}, on average, the $\rateApop$ returned by MK is positive even for the putatively nearly-neutral replicates, and significantly so for \textit{Bos} ($\rateApop$ in the range \DIFdelbegin \DIFdel{$0.65-0.68$ }\DIFdelend \DIFaddbegin \DIFadd{0.65-0.68 }\DIFaddend for genes and site) and \textit{Ovis} ($\rateApop$ in the range \DIFdelbegin \DIFdel{$0.66-0.84$ }\DIFdelend \DIFaddbegin \DIFadd{0.66-0.84 }\DIFaddend for genes and sites).
    This suggests the presence of a background of positive selection captured by MK methods but not by phylogenetic methods.
    This background signal could correspond either to adaptation specifically present in the terminal lineages on which the MK method is applied and absent over the rest of the mammalian tree, or to low-intensity recurrent positive selection, present over the tree but nevertheless missed by phylogenetic methods, owing to a lack of sensitivity.
    Alternatively, part of it could be an artifact of MK methods, due for example to a recent demographic expansion (\textit{Bos} and \textit{Ovis} are the two among those analysed by the population-genetic approach showing the highest levels of synonymous diversity), or to a more general mismatch between short- and long-term effective population size ($\Ne$)\cite{rousselle_overestimation_2018}.

    Regardless of its exact cause, subtracting this background, so as to compare, not directly the $\rateApop$ of the population-genetic method, but the $\Delta \rateApop$ between the putatively adaptive set and the control replicates, to the $\rateAphy= \omega - \omega_0$ returned by the phylogenetic method, may give a more meaningful basis for a quantitative comparison between phylogenetic and population-genetic approaches (fig.~\ref{fig:method}).
    Of note, across all analyses shown in fig.~\ref{fig:unfolded-MK}A and B, this population-genetic $\Delta \rateApop$ is always smaller than the phylogenetic $\rateAphy$.
    This asymmetry is expected, as a result of a selection bias: the genes of the test set were selected precisely for their high phylogenetic signal, while keeping a blind eye to their population-genetic signal.
    From this perspective, the ratio $\Delta \rateApop /  \rateAphy$ can be interpreted as an estimate of the fraction of the total signal captured by the phylogenetic enrichment procedure that is confirmed by MK statistics.
    This ratio, hereafter called the confirmation rate, is indicated in table~\ref{table:unfolded-MK}.

    At the gene level, the confirmation rate is relatively high, ranging from 30\% to up to 90\%.
    At the site level, the confirmation rate is lower (30\% on average), which could betray a higher rate of false discovery at the site level, or could be the result of subtle molecular evolutionary processes, such as intermittent adaptation (on some but not on all branches) or within-gene turnover (ongoing adaptation targeting different sites on different branches).

    After discarding sites with a mean $\omega > 1$, the remaining $29,543$ sites classified as being under an adaptive regime have $1 > \omega > \omega_{0}$ and are specifically discovered by the mutation-selection approach.
    Since their $\omega$ is less than 1, they could not be detected by classical codon models.
    This raises the question of the empirical value of these findings.
    Indeed, while mutation-selection methods are more sophisticated and may therefore have a greater sensitivity, they may also be more prone to producing false positives.
    The phylogenetic/population-genetic \DIFdelbegin \DIFdel{confrontation }\DIFdelend \DIFaddbegin \DIFadd{contrast }\DIFaddend developed here can be used to assess this important point.
    As shown in (fig.~\ref{fig:unfolded-MK}C) and table~\ref{table:unfolded-MK}, out of $29$ populations, for $17$ out of the $29$ populations that have been analysed, the confirmation rate is significantly positive ($\alpha$=0.05, Holm–Bonferroni correction), and of the order of 10\% on average.
    This importantly suggests the presence of a background of low-intensity positive selection, which is missed by classical codon models, but partially detected by mutation-selection models.
    In other words, the approach can detect a long-term evolutionary Red-Queen even for a site with $\omega < 1$ that is still under adaptation at the population-genetic scale.

    Because genes and sites classified as adaptive have a higher $\omega$ than genes/sites classified as nearly-neutral, $\rateApop$ could simply be higher for genes with higher $\omega$ due to this confounding factor.
    Thus we performed additional experiments where $\omega$ is controlled to be the same in the nearly-neutral replicate and the adaptive set of genes (fig.~S3-7 and tables~S5-9).
    Additionally, we performed the same experiments with a more stringent risk $\alpha=0.005$ (10 times greater) to classify genes and sites as adaptive (fig.~S8-10 and tables~S9-10).
    Our result are robust to both controlling for $\omega$ and with a different threshold to classify genes and sites as adaptive.
    Finally, we computed $\rateApop$ using the software polyDFE\cite{tataru_polydfe_2020}, which relies on the synonymous and non-synonymous unfolded site-frequency spectra (SFS) to estimate the distribution of fitness effects of mutations (DFE), and the rate of adaptation (fig.~S11-18 and tables~S11-18).
    Depending on the underlying assumptions for the shape of the DFE and the definition of $\rateApop$, we observed a wide range of $\rateApop$ both for the set of adaptive and nearly-neutral genes/sites.
    However, the statistical test for the enrichment of $\rateApop$ between the set of adaptive and nearly-neutral genes/sites gives results in the same direction whether computed by polyDFE or as \textcite{mcdonald_adaptative_1991} statistic, although the confirmation rate and the associated $p_{\mathrm{v}}$ are different.

    \section*{Discussion}\label{sec:discussion}

    Quantifying the rate of adaptation assumes that we can measure the rate of evolution and more importantly its deviation from a null model of evolution disallowing adaptation.
    For phylogenetic codon models, this null model of evolution is usually assumed to be neutral evolution and the rate of evolution computed as the ratio of non-synonymous over synonymous substitution rates ($\omega$) is thus compared to 1.
    We first showed that, at the phylogenetic scale, $\omega$ can be compared to it's expectation under the mutation-selection model ($\omega_{0}$), a nearly-neutral model instead of a neutral model of evolution, giving a quantitative estimate of the rate of adaptation as $\rateAphy = \omega - \omega_{0}$.
    Moreover, because $\omega_{0}$ is by definition lower than 1\cite{spielman_relationship_2015}, detecting adaptation as $\omega > \omega_{0}$ instead of $\omega > 1$ is always a more sensitive statistical test.
    We thus argue that mutation-selection provides both a theoretical and practical improvement over the current method of testing to detect adaptation.
    The application of this approach exome-wide across placental mammals suggests that $822$ out of $14,509$ proteins are under a long-term evolutionary Red-Queen, with ontology terms related to immune processes and the external membrane of cells.
    Enrichment in ontologies related to immune processes is expected, as found in previous studies\cite{kosiol_patterns_2008}.
    However, we also detect an enrichment in ontologies related to the external membrane and cell adhesion.
    Genes pertaining to these ontologies might be under adaptation as a consequence of being the target of viruses and parasites\cite{enard_viruses_2016, ebel_high_2017}.
    Altogether, the mutation-selection method effectively detects adaptation regardless of the background of purifying selection, and returns reasonable candidates for adaptive evolution.
    Of note, in its current implementation, and unlike classical codon models\cite{wong_accuracy_2004, yang_paml_2007}, the mutation-selection approach does not yet provide a proper and well-calibrated statistical test for calling genes or sites under adaptation with a well-controlled frequentist risk.
    This was not a problem in the enrichment analysis conducted in this article, which relies on downstream controls based on random permutations.
    Nevertheless, the encouraging results obtained here give a motivation for developing such a test, which should then have an increased power to detect adaptation, compared to classical codon models relying on the $\omega > 1$ criterion.

    At the population-genetic scale, the availability of approaches to detect adaptation\cite{mcdonald_adaptative_1991, messer_frequent_2013} raises the question whether the rate of adaptation calculated at the phylogenetic scale as $\rateAphy$ is congruent with the rate calculated at the population genetics scale by McDonald \& Kreitman (MK)\cite{mcdonald_adaptative_1991} as $\rateApop = \dnds - \pnps$.
    In this light, the set of genes and sites detected to be under adaptation at the phylogenetic scale showed a significant increase in $\rateApop$ such as inferred by population-based method (29 populations across 7 genera).
    Quantitatively, about 30\% to 90\% of the signal detected by the phylogeny-based approach is confirmed by the population-based approach.
    This result is in stark contrast with studies comparing $\omega$-based codon models at the gene level with MK methods, which found that the set of genes detected at different scales does not seem to overlap beyond random expectations\cite{chen_two_2021}.
    The reasons for this discrepancy are not totally clear.
    The use of different codon modeling strategies could play a role here.
    More fundamentally however, our study relies on a large and densely sampled phylogeny with $\simeq 100$ taxa across placental mammals, versus 5 \textit{Drosophila} and 5 \textit{Brassicaceae} in \textcite{chen_two_2021}.
    As a result, the phylogenetic aspect of our analysis benefits from an increased power, while being also inherently more focussed on genes characterized by recurrent adaptation over a very large evolutionary scale (i.e.~long-term evolutionary Red Queens), for which population-genetic signals of adaptation may be more easily recovered.
    We thus showed empirically that the mutation-selection codon model provides a null (nearly-neutral) model from which we can disentangle purifying and adaptive evolution.
    However, our procedure still has some limitations.

    First, synonymous mutations are considered neutral in our implementation of mutation-selection codon models.
    In other words, we assumed no synonymous codon usage bias (CUB), or a weak effect as it is suggested in mammals\cite{plotkin_synonymous_2011}.
    However, a strong CUB can influence the comparison between $\omega$ and $\omega_0$\cite{spielman_relationship_2015}.
    In our mutation-selection codon model, in addition to \DIFdelbegin \DIFdel{model selection for amino-acids }\DIFdelend \DIFaddbegin \DIFadd{modeling selection for amino acids }\DIFaddend at the site level, modeling CUB is possible by introducing another layer of selection for synonymous codons at the gene level\cite{yang_mutationselection_2008, rodrigue_mechanistic_2010}.
    Such a model could be used to detect adaptation in a more general context for which CUB cannot be overlooked (e.g.~\textit{Drosophila}).
    Further work will thus be required to relax the assumption that synonymous mutations are neutral and to assess the sensitivity of $\rateAphy$ to CUB.
    Second, mutation-selection codon models assume a constant effective population size while it has been established that its fluctuations has a major effect on selection dynamics\cite{lanfear_population_2014, platt_protein_2018}.
    Estimating changes in effective population size in a mutation-selection framework is possible\cite{latrille_inferring_2021}, although too computational intensive in its current implementation to be performed genome-wide.
    Third, epistasis is not modeled while it can have a large effect on the response of the rate of evolution with change in population size\cite{latrille_quantifying_2021}.
    More generally, pervasive epistasis generates an entrenchment of the amino acids\cite{goldstein_evolutionary_2004, goldstein_nonadaptive_2015, goldstein_sequence_2017}, resulting in a slowing down of the rate \DIFaddbegin \DIFadd{of }\DIFaddend evolution\cite{rodrigue_detecting_2017, patel_epistasis_2022} or a standstill\cite{youssef_evolution_2022}.
    Consequently, our estimation of the predicted rate of evolution computed at mutation-selection balance ($\omega_0$) is over-estimated given that epistasis is not taken into account in our model.
    Such an over-estimated $\omega_0$ results in an underestimated rate of adaptation $\rateAphy = \omega - \omega_0$, ultimately leading to a conservative test of adaptation which could theoretically be improved.
    Accurately modeling the underlying fitness landscape is thus a direction for future research in mutation-selection models to improve the estimation of $\rateAphy$, by means of relaxing model assumptions\cite{goldstein_sequence_2017} or by leveraging empirically determined fitness landscapes\cite{bloom_identification_2017, schrempf_scalable_2020}.

    On the other hand, at the population-genetic scale, the greatest limitation to detecting adaptation is the lack of power determined by the genetic diversity since polymorphisms are rare and estimation of $\pnps$ requires to pool many sites for which SNPs are available.
    Since the effects of mild purifying selection are more pronounced on longer time scales (i.e.~mildly deleterious mutations contribute disproportionately to polymorphism, compared to divergence), $\rateApop$ as computed by MK can be biased by moderately deleterious mutations\cite{eyre-walker_quantifying_2002, ho_time_2005} and by the change in population size through time\cite{eyre-walker_changing_2002}.
    To overcome this bias, model-based approaches relying on the synonymous and non-synonymous site-frequency spectra (SFS) to estimate the distribution of fitness effects of mutations (DFE), so as to account for the contribution of mild selective effects to standing polymorphism, have been developed\cite{eyre-walker_estimating_2009, tataru_polydfe_2020} and are often used\cite{galtier_adaptive_2016, moutinho_variation_2019}.
    However, the broad range of $\rateApop$ estimated on sets of genes/sites classified as nearly-neutral suggests that these models are lacking power, even more than the MK statistic, because of the sparsity of the SFS.
    Beside changes in population size biasing the estimation\cite{rousselle_overestimation_2018}, we argue that inferring $\rateApop$ using an underlying DFE model is also highly sensitive to assumptions for the shape of the DFE and the definition of $\rateApop$.
    For example, the value of $\rateApop$ is computed as an integral, where the bounds of this integral is debated by different authors\cite{galtier_adaptive_2016, tataru_polydfev2_2019}.
    It is thus relatively easy to change the definition of $\rateApop$ (fig.~S13-16 and tables~S13-16) or to constrain the underlying DFE (fig.~S13-18 and tables~S13-18) to obtain a wide range of $\rateApop$ on the same dataset.
    Taken together, we argue that comparing $\rateApop$ to 0 is not a robust test for adaptation.
    Instead, $\rateApop$ for a particular genomic region of interest should be compared to other genomic regions for which the nearly-neutral evolution is not rejected, and the difference $\Delta \rateApop$ should be compared to 0, as done in this study.
    More generally, our empirical analysis emphasizes the limitations of, and the difficulties raised by, the model-based population-genetic approaches.
    In this respect, further exploring the congruence (or lack thereof) between phylogenetic and population-genetic approaches will represent a useful asset to clarify those delicate problems, given that similar benefits are also expected on the side of phylogenetic approaches, which are far from immune from methodological limitations.

    More broadly on a theoretical level, this work leverages a specific overlap between phylogenetic and population genetics, namely that the rate of adaptation $\rateAphy$ in phylogenetic codon models and $\rateApop$ in the MK test should theoretically be directly comparable.
    Based on this theoretical relationship, our study is paving the way for studies and methods augmenting molecular polymorphism data within species with information about divergence data between species\cite{chen_hunting_2021}, and by assessing empirically the relationship between phylogenetic and population genetics \DIFaddbegin \DIFadd{analyses}\DIFaddend \cite{thorne_codon_2012}.
    In this light, mutation-selection models at the phylogenetic scale can play a dual role: pinpointing genes and sites under adaptation ($\rateAphy > 0$), and also seeking the genomic region for which the nearly-neutral theory is not rejected ($\rateAphy \simeq 0$).

    \section*{Methods}\label{sec:methods}

    \subsection*{Phylogenetic dataset}
    Protein-coding DNA sequences alignments in placental mammals and their corresponding gene trees were extracted from the \href{https://www.orthomam.univ-montp2.fr}{OrthoMaM} database, containing $116$ mammalian reference sequences in v10c\cite{ranwez_orthomam_2007, douzery_orthomam_2014, scornavacca_orthomam_2019}.
    Genes located on the X, Y and mitochondrial chromosome were discarded from the analysis, since the number of polymorphism, necessary in population-based method, is expected to be different on these sequences.
    Additionally, sequences from the species for which polymorphism are available, as well as their sister species have been discarded from the analysis to ensure independence between the data used in the phylogenetic and population-genetic method, \DIFdelbegin \DIFdel{such }\DIFdelend \DIFaddbegin \DIFadd{so }\DIFaddend as to avoid fallacy and circularity in the estimation of $\rateApop$ and $\rateAphy$.
    Altogether, we analyzed $14,509$ protein-coding DNA sequences alignment containing at most $87$ reference sequences of placental mammals.

    \subsection*{Adaptation in phylogeny-based method}
    Classical codon models \DIFdelbegin \DIFdel{estimates }\DIFdelend \DIFaddbegin \DIFadd{estimate }\DIFaddend a parameter $\omega=\dnds$, namely the ratio of the non-synonymous over the synonymous substitution rates\cite{muse_likelihood_1994,goldman_codonbased_1994}.
    In the so-called site models, $\omega$ is allowed to vary across sites\cite{yang_codonsubstitution_2000, huelsenbeck_dirichlet_2006}.
    In \textit{Bayescode}, site-specific $\omega^{(i)}$ (fig.~\ref{fig:scatterplot}B, y-axis) are modeled as independent and identically distributed random effects from a gamma distribution\cite{lartillot_phylobayes_2013}.
    Therefore, this corresponds to a continuous version of the M5 model of \textcite{yang_codonsubstitution_2000}.
    In a second step, the average over sites is calculated, giving estimates of $\omega$ for each protein-coding sequence (fig.~\ref{fig:scatterplot}A, y-axis).
    To assess the sensitivity of our estimate of $\omega$ to the prior as implemented in \textit{BayesCode}, we ran \textit{CODEML} site model (F1x4 MG, 8 discrete categories of $\omega$) from the PAML software\cite{yang_paml_2007} on a sample of 100 random genes.
    Both $\omega$ estimates from \textit{BayesCode} and \textit{CODEML} are strongly correlated at the site level ($r^2=0.985$, slope of $0.92$ in fig.~S1.A) and at the gene level ($r^2=0.991$, slope of $0.9$ in fig.~S1.B).

    In contrast to $\omega$-based codon models, mutation-selection models assume that the protein-coding sequence is at mutation-selection balance under a fixed fitness landscape, which is itself characterized by a fitness vector over the $20$ amino \DIFdelbegin \DIFdel{acid }\DIFdelend \DIFaddbegin \DIFadd{acids }\DIFaddend at each site\cite{yang_mutationselection_2008, halpern_evolutionary_1998, rodrigue_mechanistic_2010}.
    Mathematically, the rate of non-synonymous substitution from codon $a$ to codon $b$ ($q_{a \mapsto b}^{(i)}$) at site $i$ of the sequence is equal to the rate of mutation from the underlying DNA change ($\mu_{a \mapsto b}$) multiplied by the scaled probability of fixation of the mutation ($\proba_{a \mapsto b}^{(i)}$).
    Crucially, the probability of fixation depends on the difference of scaled fitness between the amino acid encoded by the mutated codon ($F_b^{(i)}$) and the fitness of the amino acid encoded by the original codon ($F_a^{(i)}$) \DIFdelbegin \DIFdel{of }\DIFdelend \DIFaddbegin \DIFadd{at a given }\DIFaddend site $i$\DIFdelbegin \DIFdel{\mbox{%DIFAUXCMD
\cite{wright_evolution_1931, fisher_genetical_1930}}\hspace{0pt}%DIFAUXCMD
}\DIFdelend \DIFaddbegin \DIFadd{\mbox{%DIFAUXCMD
\cite{fisher_distribution_1930, wright_evolution_1931, mccandlish_modeling_2014}}\hspace{0pt}%DIFAUXCMD
}\DIFaddend .
    Altogether, the rate of substitution from codon $a$ to $b$ at a given site $i$ is:
    \begin{equation}
        q_{a \mapsto b}^{(i)} = \mu_{a \mapsto b} \proba_{a \mapsto b}^{(i)} = \mu_{a \mapsto b} \dfrac{F_b^{(i)} - F_a^{(i)}}{1 - \e^{F_a^{(i)} - F_b^{(i)}}}.
    \end{equation}

    Fitting the mutation-selection model on a sequence alignment leads to an estimation of the mutation rate matrix ($\UniDimArray{\mu}$) as well as the 20 \DIFdelbegin \DIFdel{amino acid }\DIFdelend \DIFaddbegin \DIFadd{amino-acid }\DIFaddend fitness landscape ($\UniDimArray{F^{(i)}}$) at each site $i$.
    From these parameters, one can compute $\omega_{0}^{(i)}$ (fig.~\ref{fig:scatterplot}B, x-axis), the site-specific rate of non-synonymous over synonymous substitution at the mutation-selection balance:
    \begin{equation}
        \omega_{0}^{(i)} = \dfrac{\sum_{a \in \mathcal{C}} \sum_{b \in \mathcal{N}_a} \pi_a^{(i)} q_{a \mapsto b}^{(i)}}{\sum_{a \in \mathcal{C}} \sum_{b \in \mathcal{N}_a} \pi_a^{(i)} \mu_{a \mapsto b}},
    \end{equation}
    where $\mathcal{C}$ is the set all the possible codons ($61$ by discarding stop codons), $\pi_a^{(i)}$ is the equilibrium frequency of codon $a$ at site $i$, and $\mathcal{N}_a$ is the set of codons that are non-synonymous to $a$\cite{spielman_relationship_2015, rodrigue_detecting_2017}.
    The equilibrium frequency of codon $a$ at site $i$ is \DIFdelbegin \DIFdel{the product }\DIFdelend \DIFaddbegin \DIFadd{a function }\DIFaddend of the nucleotide frequencies at its three positions and the \DIFdelbegin \DIFdel{scaled Wrightian fitness of the amino acid }\DIFdelend \DIFaddbegin \DIFadd{amino-acid scaled fitness }\DIFaddend ($F_a^{(i)}$):
    \begin{align}
        \pi_a^{(i)} & = \dfrac{ \sigma_{a[1]}\sigma_{a[2]}\sigma_{a[3]} \e^{F_a^{(i)}}}{\sum\limits_{b=1}^{61}\sigma_{b[1]}\sigma_{b[2]}\sigma_{b[3]} \e^{F_b^{(i)}} },
    \end{align}
    where $\sigma_{a[j]} \in \{A, T, C, G\}$ is the equilibrium frequency (given by the mutational matrix) of the nucletoide at position $j \in \{1, 2, 3\}$ of codon $a$.
    In a second step, the average over sites is calculated, giving estimates of $\omega_{0}$ for each protein-coding sequences (fig.~\ref{fig:scatterplot}A, x-axis).
    Under the assumption that the protein is under a nearly-neutral regime, the calculated $\omega_{0}$ (mutation-selection model) and the estimated $\omega$ (site model) should be the same\cite{spielman_relationship_2015}.

    We ran the Bayesian software \textit{BayesCode} (\url{https://github.com/ThibaultLatrille/bayescode}) on each protein-coding DNA alignment\cite{rodrigue_bayesian_2021}.
    Each Monte-Carlo Markov-Chain (MCMC) is run during $2,000$ points, with a burn-in of $1,000$ points, to obtain the posterior mean of $\omega$ and $\omega_{0}$ across the MCMC, as well as the $95$\% posterior credibility interval for genes and sites.
    Genes and sites classified under an adaptive regime (in red) are rejecting the nearly-neutral assumption such that the lower bound for the credible interval of $\omega$ ($\alpha=0.05$) is above the upper bound of the credible interval of $\omega_{0}$ ($\alpha=0.05$), meaning that the value of their $\omega$ is higher than that of their $\omega_{0}$.
    Because this is a unilateral test ($\omega > \omega_{0}$) and the two credible interval are independent, the risk is $(\alpha/2)^2=0.025^2=0.000625$ for each test.
    Empirically, the nearly-neutral assumption appears to be rejected for $822$ out \DIFaddbegin \DIFadd{of }\DIFaddend $14,509$ genes, while $0.000625 \times 14,509 \simeq 9$ genes are expected due to the multiple testing, suggesting a $9 / 822 \simeq 1\%$ rate of false positive at the gene level.
    At the site level, the nearly-neutral assumption appears to be rejected for $104,129$ out of $8,895,374$ sites, while $0.000625 \times 8,895,374 \simeq 5,560$ are expected due to the multiple testing, suggesting a $5,560 / 104,129 \simeq 5\%$ rate of false positive at the site level.
    Genes and sites are classified under a nearly-neutral regime (in green) if the average $\omega$ is within the credible interval of the $\omega_{0}$, and respectively the average $\omega_{0}$ is also within the credible interval of  $\omega$, meaning $\omega = \omega_{0}$.
    Additionally, the set of sites detected exclusively by mutation-selection codon models have a mean $\omega < 1 $.
    Genes and sites that do not fall in any of these categories are considered unclassified.

    \subsection*{Polymorphism dataset}

    Each SNP (chromosome, position, strand) in the focal species was matched to its relative position (chromosome, position, strand) in the protein-coding DNA alignment by first converting the genomic positions to relative position in the coding sequence (CDS) using gene annotation files (GTF format) downloaded from Ensembl (\url{ensembl.org}).
    We then verified that the SNP downloaded from Ensembl were matching the reference in the CDS (FASTA format).
    Second, the relative position in the CDS was converted to position in the multiple sequence alignment (containing gaps) from OrthoMaM database\cite{ranwez_orthomam_2007, douzery_orthomam_2014, scornavacca_orthomam_2019} by doing a global pairwise alignment, using the Biopython function pairwise2, between the CDS fasta and the sequence found in the alignment.
    This conversion from genomic position to position in the alignment is only possible if the assembly used for SNP calling is the same as the one used in the alignment, the GTF annotations and the FASTA sequences.

    We retrieved the genetic variants representing the population level polymorphism from the following species and respective available datasets: \textit{Equus caballus} (EquCab2 assembly in the EVA study PRJEB9799\cite{alabri_whole_2020}), \textit{Canis familiars} (CanFam3.1 assembly in the EVA study PRJEB24066\cite{jagannathan_comprehensive_2019}),  \textit{Bos taurus} (UMD3.1 assembly in the NextGen project), \textit{Ovis aries} (Oar\_v3.1 assembly in the NextGen project), \textit{Capra Hircus} (CHIR1 assembly in the NextGen project converted to ARS1 assembly with dbSNP identifiers\cite{sherry_dbsnp_2001}), \textit{Chlorocebus sabaeus} (ChlSab1.1 assembly in the EVA project PRJEB22989\cite{svardal_ancient_2017}), \textit{Homo sapiens} (GRCh38 assembly from the 1000-genome project\cite{consortium_integrated_2012, the1000genomesprojectconsortium_global_2015}).

    Variants not inside genes are discarded at the beginning of the analysis.
    Insertions and deletions are not analyzed, and only Single Nucleotide Polymorphisms (SNPs) with only one mutant allele are considered.
    Stop codon mutants are also discarded.
    For populations containing more than $8$ sampled individuals, the site-frequency spectrum (SFS) is subsampled down to $16$ chromosomes ($8$ diploid individuals) without replacement \DIFdelbegin \DIFdel{(hyper-geometric distribution) }\DIFdelend to alleviate the effect of different sampling depth in the $29$ populations.
    Moreover, subsampling mitigate the impact of moderately deleterious mutations segregating at low frequency on $\pnps$, since they are more likely to be discarded than polymorphism segregating at higher frequency.
    The Snakemake pipeline for integrating polymorphism and divergence data uses custom scripts written in python 3.9.

    \subsection*{Rate of adaption in population-based method}

    The genes and sites classified as under adaptation are concatenated.
    For each population $\pnps$ is computed as the sum of non-synonymous over synonymous polymorphism on the concatenated SFS\@.
    $\dnds$ is computed on the concatenated pairwise alignment between focal and sister species extracted from OrthoMaM, the $\dnds$ count is performed by \textit{yn00}.
    We considered \textit{Ceratotherium simum simum} as \textit{Equus caballus} sister species; \textit{Ursus maritimus} as \textit{Canis familiars} sister species; \textit{Bison bison bison} as \textit{Bos taurus} sister species; \textit{Pantholops hodgsonii} as \textit{Ovis aries} sister species; \textit{Pantholops hodgsonii} as \textit{Capra Hircus} sister species; \textit{Macaca mulatta} as \textit{Chlorocebus sabaeus} sister species and finally we considered \textit{Pan troglodytes} as \textit{Homo sapiens} sister species.
    Altogether, $\rateApop = \dnds - \pnps$ is thus computed for each population on genes and sites classified as under adaptation.
    The result is compared to the empirical null distribution of $\rateApop$, obtained by randomly sampling ($1,000$ sampling replicates) a subset of genes/sites classified as nearly-neutral.

    Other methods to compute $\rateApop$ such as polyDFE\cite{tataru_polydfe_2020} are also used (eq.~3-20 in supplementary materials), which relies on the synonymous and non-synonymous unfolded site-frequency spectra (SFS) to estimate the distribution of fitness effects of mutations (DFE), and the rate of adaptation.
    In polyDFE, GammaExpo models the fitness effect of weakly deleterious non-synonymous mutations as distributed according to a negative Gamma and the fitness effect of weakly advantageous mutations are distributed exponentially.
    This method is an extension of the methods introduced by Eyre-Walker and collaborators\cite{eyre-walker_distribution_2006, eyre-walker_estimating_2009}.
    Unfolded SFSs are obtained by polarizing SNPs using the $3$ closest outgroups found in the OrthoMam alignment with est-usfs v2.04\cite{keightley_inferring_2018}.

    \section{Data availability}\label{sec:data-availability}
    The data underlying this article are available at \url{https://doi.org/10.5281/zenodo.7107233}.
    Scripts and instructions necessary to reproduce the empirical experiments on the original dataset or with user-specified datasets is available at \url{https://github.com/ThibaultLatrille/AdaptaPop}.

    \section{Acknowledgements}\label{sec:acknowledgements}
    We gratefully also acknowledge the help of Nicolas Galtier and Julien Joseph for their advice and review concerning this manuscript.
    This work was performed using the computing facilities of the CC LBBE/PRABI.
    This study makes use of data generated by the NextGen Consortium.
    The European Union’s Seventh Framework Programme (FP7/2010-2014) provided funding for the project under grant agreement no 244356 - “NextGen”.
    Funding: Université de Lausanne.
    Agence Nationale de la Recherche, Grant ANR-15-CE12-0010-01 / DASIRE.
    Agence Nationale de la Recherche, Grant ANR-19-CE12-0019 / HotRec.
    Natural Sciences and Engineering Research Council of Canada.

%TC:ignore
    \section{Author information}\label{sec:author-information}
    TL, NR and NL designed the study.
    TL gathered and formatted the data and conducted the analyses with \texttt{BayesCode} using scripts in \texttt{Python} and pipeline in \texttt{Snakemake}.
    TL, NR and NL contributed to the writing of the manuscript.

    \printbibliography
%TC:endignore
    \begin{figure*}[htb]
        \centering
        \includegraphics[width=0.95\linewidth, page=1]{figure1}
        \caption{
            Detection of protein-coding sequences ongoing adaptation at the phylogenetic scale.
            $\omega$ estimated by the site model against $\omega_{0}$ calculated by the mutation-selection model.
            Scatter plot of $14,509$ genes in panel A, with $95$\% bayesian credible interval ($\alpha=0.05$).
            Density plot of sites in panel B and C.
            Genes or sites are then classified whether they detected as adaptive ($\omega > \omega_{0}$ in red) or nearly-neutral ($\omega \simeq \omega_{0}$ in green).
            In panel C, the set of sites detected exclusively by mutation-selection codon models have a mean $\omega < 1 $.}
        \label{fig:scatterplot}
    \end{figure*}

    \begin{figure*}[htb]
        \centering
        \includegraphics[width=0.95\linewidth, page=1]{figure2}
        \caption{
            Integrating divergence and polymorphism for the detection of adaptation.
            At the phylogenetic level, $\omega$ (classical codon models) and $\omega_{0}$ (mutation-selection codon models) are computed from protein-coding DNA alignments, allowing \DIFaddbeginFL \DIFaddFL{us }\DIFaddendFL to classify genes into adaptive (in red) and nearly-neutral (in green) regime.
            At the population-genetic level, for each population, $\rateApop$ is computed on the \DIFdelbeginFL \DIFdelFL{concatenate }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{concatenated sequence }\DIFaddendFL of genes classified as under adaptation.
            The result is compared to the empirical null distribution of $\rateApop$ in each population, obtained by randomly sampling ($1,000$ replicates) a subset under a nearly-neutral regime.
        }
        \label{fig:method}
    \end{figure*}

    \begin{figure*}[htb]
        \centering
        \includegraphics[width=0.95\linewidth, page=1]{figure3}
        \caption{
            Enrichment of adaptation at the population-genetic scale for $29$ populations across $7$ genera at the gene (panel A) and site (panels B and C) level.
            For each population, $\rateApop$ is computed on $822$ genes (A) and $104,129$ sites (B) having a high rate of adaptation at the phylogenetic scale ($\omega > \omega_0$ in red).
            In panel C, the set of 29,543 sites are detected exclusively by mutation-selection codon models with a mean $\omega < 1 $.
            The result is compared to the empirical null distribution of $\rateApop$, obtained by randomly sampling ($1,000$ replicates) a subset of genes and sites under a nearly-neutral regime (violin plot in green).
            $^*$ signify that the $p_{\mathrm{v}}$ corrected for multiple comparison (Holm–Bonferroni correction) is lower than the risk $\alpha=0.05$.
            The acronym of populations, and the quantitative value of $\rateApop$ and $p_{\mathrm{v}}$ are shown in table~\ref{table:unfolded-MK}}
        \label{fig:unfolded-MK}
    \end{figure*}

    \begin{table*}[tb]
        \centering
        \begin{adjustbox}{width = 0.95\textwidth}
            \begin{tabular}{||l|l|r||r|r|r||r|r|r||r|r|r||}
                \toprule
                \multicolumn{3}{||c||}{} &
                \multicolumn{3}{c||}{\textbf{Genes (822)}} &
                \multicolumn{3}{c||}{\textbf{Sites (104,129)}} &
                \multicolumn{3}{c||}{\textbf{Sites ($\bm{\omega < 1}$) (29,543)}}
                \\ \hline
                \textbf{Population} &
                \textbf{Species} &
                $\bm{\pi_{\textrm{S}}}$ &
                $\bm{\Delta \rateApop}$ &
                $\bm{p_{\mathrm{v}}^{\mathrm{adj}}}$    &
                $\bm{\frac{\Delta\rateApop}{\rateAphy}}$ &
                $\bm{\Delta \rateApop}$ &
                $\bm{p_{\mathrm{v}}^{\mathrm{adj}}}$          &
                $\bm{\frac{\Delta\rateApop}{\rateAphy}}$ &
                $\bm{\Delta \rateApop}$ &
                $\bm{p_{\mathrm{v}}^{\mathrm{adj}}}$ &
                $\bm{\frac{\Delta\rateApop}{\rateAphy}}$
                \\                \midrule
                Diverse (Equus)                                     & Equus caballus      & $ 0.002$ & $ 0.094$ & $\bm{0.0{^*}}$    & $ 0.928$ & $ 0.399$ & $\bm{0.0{^*}}$ & $ 0.459$ & $ 0.258$ & $\bm{0.0{^*}}$ & $ 0.446$ \\
                \rowcolor{LIGHTGREY} Diverse (Canis)                & Canis familiaris    & $ 0.004$ & $ 0.058$ & $\bm{0.0{^*}}$                  & $ 0.557$                                                                     & $ 0.406$ & $\bm{0.0{^*}}$ & $ 0.463$ & $ 0.227$ & $\bm{0.0{^*}}$ & $ 0.392$ \\
                Iran (IRBT)                                         & Bos taurus          & $ 0.008$ & $ 0.028$ & $\bm{ 0.020{^*}}$ & $ 0.278$ & $ 0.237$ & $\bm{0.0{^*}}$ & $ 0.272$ & $ 0.134$ & $ 0.150~~$ & $ 0.231$ \\
                Uganda (UGBT)                                       & Bos taurus          & $ 0.008$ & $ 0.036$ & $\bm{0.0{^*}}$    & $ 0.355$ & $ 0.222$ & $\bm{0.0{^*}}$ & $ 0.254$ & $ 0.156$ & $\bm{ 0.017{^*}}$ & $ 0.270$ \\
                \rowcolor{LIGHTGREY} Australia (AUCH)               & Capra hircus        & $ 0.003$ & $ 0.052$ & $\bm{0.0{^*}}$                  & $ 0.506$                                                                     & $ 0.202$ & $\bm{0.0{^*}}$ & $ 0.230$ & $ 0.168$ & $ 0.143~~$ & $ 0.290$ \\
                \rowcolor{LIGHTGREY} France (FRCH)                  & Capra hircus        & $ 0.003$ & $ 0.073$ & $\bm{0.0{^*}}$                  & $ 0.709$                                                                     & $ 0.220$                      & $\bm{0.0{^*}}$ & $ 0.250$ & $ 0.236$ & $\bm{ 0.039{^*}}$ & $ 0.407$ \\
                \rowcolor{LIGHTGREY} Iran (IRCA)                    & Capra aegagrus      & $ 0.004$ & $ 0.049$ & $\bm{0.0{^*}}$                  & $ 0.482$                                                                     & $ 0.242$                      & $\bm{0.0{^*}}$ & $ 0.275$ & $ 0.108$ & $ 0.396~~$ & $ 0.186$ \\
                \rowcolor{LIGHTGREY} Iran (IRCH)                    & Capra hircus        & $ 0.004$ & $ 0.062$ & $\bm{0.0{^*}}$    & $ 0.610$                                                                     & $ 0.210$                      & $\bm{0.0{^*}}$ & $ 0.239$ & $ 0.217$ & $\bm{ 0.017{^*}}$ & $ 0.375$ \\
                \rowcolor{LIGHTGREY} Italy (ITCH)                   & Capra hircus        & $ 0.003$ & $ 0.052$ & $\bm{0.0{^*}}$    & $ 0.511$                                                                     & $ 0.174$                      & $\bm{0.0{^*}}$ & $ 0.199$ & $ 0.134$ & $ 0.308~~$ & $ 0.232$ \\
                \rowcolor{LIGHTGREY} Morocco (MOCH)                 & Capra hircus        & $ 0.004$ & $ 0.064$ & $\bm{0.0{^*}}$                  & $ 0.626$                                                                     & $ 0.256$                      & $\bm{0.0{^*}}$ & $ 0.292$ & $ 0.201$ & $\bm{0.0{^*}}$ & $ 0.347$ \\
                Iran (IROA)                                         & Ovis aries          & $ 0.007$ & $ 0.087$ & $\bm{0.0{^*}}$    & $ 0.847$ & $ 0.199$ & $\bm{0.0{^*}}$ & $ 0.228$ & $ 0.183$ & $\bm{ 0.017{^*}}$ & $ 0.316$ \\
                Iran (IROO)                                         & Ovis orientalis     & $ 0.009$ & $ 0.087$ & $\bm{0.0{^*}}$    & $ 0.848$ & $ 0.204$ & $\bm{0.0{^*}}$ & $ 0.233$ & $ 0.176$ & $\bm{0.0{^*}}$ & $ 0.304$ \\
                Iran (IROV)                                         & Ovis vignei         & $ 0.005$ & $ 0.072$ & $\bm{0.0{^*}}$    & $ 0.697$ & $ 0.194$ & $\bm{0.0{^*}}$ & $ 0.222$ & $ 0.192$ & $\bm{0.0{^*}}$ & $ 0.332$ \\
                Various (ISGC)                                      & Ovis aries          & $ 0.008$ & $ 0.076$ & $\bm{0.0{^*}}$    & $ 0.742$ & $ 0.171$ & $\bm{0.0{^*}}$ & $ 0.195$ & $ 0.189$ & $\bm{0.0{^*}}$ & $ 0.326$ \\
                Morocco (MOOA)                                      & Ovis aries          & $ 0.008$ & $ 0.093$ & $\bm{0.0{^*}}$    & $ 0.905$ & $ 0.189$ & $\bm{0.0{^*}}$ & $ 0.216$ & $ 0.193$ & $\bm{0.0{^*}}$ & $ 0.333$ \\
                \rowcolor{LIGHTGREY} Barbados                       & Chlorocebus sabaeus & $ 0.003$ & $ 0.068$ & $\bm{0.0{^*}}$                  & $ 0.665$                                                                     & $ 0.341$                      & $\bm{0.0{^*}}$ & $ 0.390$ & $ 0.248$ & $\bm{0.0{^*}}$ & $ 0.430$ \\
                \rowcolor{LIGHTGREY} Central African Republic (CAR) & Chlorocebus sabaeus & $ 0.006$           & $ 0.034$                      & $\bm{0.0{^*}}$ & $ 0.334$ & $ 0.229$ & $\bm{0.0{^*}}$ & $ 0.262$ & $ 0.195$ & $\bm{0.0{^*}}$ & $ 0.338$ \\
                \rowcolor{LIGHTGREY} Ethiopia                       & Chlorocebus sabaeus & $ 0.005$ & $ 0.044$ & $\bm{0.0{^*}}$                  & $ 0.425$                                                                     & $ 0.231$                      & $\bm{0.0{^*}}$ & $ 0.264$ & $ 0.264$ & $\bm{0.0{^*}}$ & $ 0.457$ \\
                \rowcolor{LIGHTGREY} Gambia                         & Chlorocebus sabaeus & $ 0.005$ & $ 0.041$ & $\bm{0.0{^*}}$                  & $ 0.403$                                                                     & $ 0.236$                      & $\bm{0.0{^*}}$ & $ 0.270$ & $ 0.217$ & $\bm{0.0{^*}}$ & $ 0.375$ \\
                \rowcolor{LIGHTGREY} Kenya                          & Chlorocebus sabaeus & $ 0.004$ & $ 0.061$ & $\bm{0.0{^*}}$    & $ 0.598$                                                                     & $ 0.181$                      & $\bm{0.0{^*}}$ & $ 0.207$ & $ 0.152$ & $ 0.150~~$ & $ 0.264$ \\
                \rowcolor{LIGHTGREY} Nevis                          & Chlorocebus sabaeus & $ 0.003$ & $ 0.029$ & $\bm{ 0.020{^*}}$               & $ 0.279$                                                                     & $ 0.332$                      & $\bm{0.0{^*}}$ & $ 0.380$ & $ 0.237$ & $\bm{ 0.017{^*}}$ & $ 0.410$ \\
                \rowcolor{LIGHTGREY} South Africa (SA)              & Chlorocebus sabaeus & $ 0.006$ & $ 0.065$ & $\bm{0.0{^*}}$                  & $ 0.633$                                                                     & $ 0.199$ & $\bm{0.0{^*}}$ & $ 0.228$ & $ 0.142$ & $ 0.108~~$ & $ 0.246$ \\
                \rowcolor{LIGHTGREY} Saint Kitts (SK)               & Chlorocebus sabaeus & $ 0.004$ & $ 0.040$ & $\bm{0.0{^*}}$                  & $ 0.388$                                                                     & $ 0.324$ & $\bm{0.0{^*}}$ & $ 0.371$ & $ 0.253$ & $\bm{0.0{^*}}$ & $ 0.439$ \\
                \rowcolor{LIGHTGREY} Zambia                         & Chlorocebus sabaeus & $ 0.006$ & $ 0.066$ & $\bm{0.0{^*}}$                  & $ 0.642$                                                                     & $ 0.132$                      & $\bm{0.0{^*}}$ & $ 0.151$ & $ 0.131$ & $ 0.150~~$ & $ 0.227$ \\
                African (AFR)                                       & Homo sapiens        & $ 0.002$ & $ 0.059$ & $\bm{ 0.012{^*}}$ & $ 0.568$ & $-0.010$ & $ 1.000~~$ & $-0.012$ & $ 0.089$ & $ 1.000~~$ & $ 0.155$ \\
                Ad Mixed American (AMR)                             & Homo sapiens        & $ 0.002$ & $ 0.067$ & $\bm{ 0.006{^*}}$ & $ 0.647$ & $-0.029$ & $ 1.000~~$ & $-0.034$ & $-0.141$ & $ 1.000~~$ & $-0.244$ \\
                East Asian (EAS)                                    & Homo sapiens        & $ 0.002$ & $ 0.063$ & $\bm{ 0.006{^*}}$ & $ 0.610$ & $-0.096$ & $ 1.000~~$ & $-0.111$ & $-0.296$ & $ 1.000~~$ & $-0.513$ \\
                European (EUR)                                      & Homo sapiens        & $ 0.002$ & $ 0.061$ & $\bm{ 0.015{^*}}$ & $ 0.590$ & $-0.078$ & $ 1.000~~$ & $-0.089$ & $-0.289$ & $ 1.000~~$ & $-0.500$ \\
                South Asian (SAS)                                   & Homo sapiens        & $ 0.002$ & $ 0.089$ & $\bm{0.0{^*}}$    & $ 0.866$ & $-0.113$ & $ 1.000~~$ & $-0.130$ & $-0.111$ & $ 1.000~~$ & $-0.193$ \\
                \bottomrule
            \end{tabular}
        \end{adjustbox}
        \caption{
            Across $29$ populations (rows), table of quantitative value of $\Delta \rateApop$ between the set classified as adaptive and nearly-neutral shown in fig.~\ref{fig:unfolded-MK}.
            $p_{\mathrm{v}}^{\mathrm{adj}}$ associated to the test are corrected for multiple comparison (Holm–Bonferroni correction, $^*$ for $p_{\mathrm{v}}^{\mathrm{adj}} < 0.05$).
            $\frac{\Delta\rateApop}{\rateAphy}$ is the ratio of $\Delta \rateApop$ at the population-genetic level and the phylogenetic level.
            $\pi_{\textrm{S}}$ is the observed genetic diversity (number of SNPs per site) counted over synonymous sites.
        }
        \label{table:unfolded-MK}
    \end{table*}
\end{document}
