%! BibTeX Compiler = biber
\documentclass{article}

\usepackage{xcolor, colortbl}
\definecolor{BLUELINK}{HTML}{0645AD}
\definecolor{DARKBLUELINK}{HTML}{0B0080}
\definecolor{LIGHTBLUELINK}{HTML}{3366BB}
\definecolor{PURPLELINK}{HTML}{663366}
\definecolor{RED}{HTML}{EB6231}
\definecolor{GREEN}{HTML}{8FB03E}
\definecolor{LIGHTGREY}{gray}{0.9}
\PassOptionsToPackage{hyphens}{url}
\usepackage[colorlinks=false]{hyperref}
% for linking between references, figures, TOC, etc in the pdf document
\hypersetup{colorlinks,
	linkcolor=DARKBLUELINK,
	anchorcolor=DARKBLUELINK,
	citecolor=DARKBLUELINK,
	filecolor=DARKBLUELINK,
	menucolor=DARKBLUELINK,
	urlcolor=BLUELINK
} % Color citation links in purple
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{naturalnames}{hyperref}

\usepackage{biorxiv}
\usepackage[backend=biber,style=nature]{biblatex}
\addbibresource{codon_models.bib}

\usepackage{url}
\usepackage{amssymb,amsfonts,amsmath,amsthm,mathtools}
\usepackage{textcomp}
\usepackage{gensymb}
\usepackage{cancel}
\usepackage{lmodern}
\usepackage{xfrac, nicefrac}
\usepackage{blkarray}
\usepackage{pgf,tikz}
\usetikzlibrary{positioning,arrows,automata,calc}
\usepackage{bm}
\usepackage{listings, enumerate, enumitem}
\usepackage[export]{adjustbox}
\usepackage{graphicx}
\usepackage{tabu}
\usepackage{hhline}
\usepackage{multicol,multirow,array}
\usepackage{etoolbox}
\usepackage{booktabs}

\usepackage[flushleft] {threeparttable}
\usepackage{bbold}
\usepackage{pdfpages}
\usetikzlibrary{positioning}
\usetikzlibrary{arrows,automata}
\pdfinclusioncopyfonts=1
\usepackage{nicefrac} % compact symbols for 1/2, etc.
\usepackage{microtype} % microtypography
\usepackage{lineno}

\graphicspath{{artworks/}}
\makeatletter
\def\input@path{{artworks/}}
\makeatother

\newcommand{\specialcell}[2][c]{%
	\begin{tabular}[#1]{@{}c@{}}#2\end{tabular}}

\newcommand{\UniDimArray}[1]{\bm{#1}}
\newcommand{\BiDimArray}[1]{\bm{#1}}

\DeclareMathOperator{\E}{\mathbb{E}}
\DeclareMathOperator{\Var}{\mathrm{Var}}
\newcommand{\der}{\mathrm{d}}
\newcommand{\angstrom}{\mathrm{\normalfont\AA}}
\newcommand{\e}{\mathrm{e}}
\newcommand{\avg}[1]{\left< #1 \right>} % for average
\newcommand{\Ne}{N_{\mathrm{e}}}
\newcommand{\Ner}{N_{\mathrm{r}}}
\newcommand{\dn}{d_N}
\newcommand{\ds}{d_S}
\newcommand{\dnds}{\dn / \ds}
\newcommand{\pn}{\pi_N}
\newcommand{\ps}{\pi_S}
\newcommand{\pnps}{\pn / \ps}
\newcommand{\proba}{\mathbb{P}}
\newcommand{\pfix}{\proba_{\mathrm{fix}}}
\newcommand{\Pfix}{2 \Ne \proba_{\mathrm{fix}}}
\newcommand{\indice}{a}
\newcommand{\indiceexp}{^{(\indice)}}

\renewcommand{\baselinestretch}{1.5}
\renewcommand{\arraystretch}{0.6}
\linenumbers

\title{Reconciling methods for detecting adaptation from inter- and intra-specific data}

%\date{September 9, 1985}	% Here you can change the date presented in the paper title
%\date{} \rowcolor{LIGHTGREY}% Or removing it

\author{
	\large
	T. {Latrille}$^{1,2}$, N. {Rodrigue}$^{3}$, N. {Lartillot}$^{1}$\\
	\normalsize
	$^{1}$Université de Lyon, Université Lyon 1, CNRS, Laboratoire de Biométrie et Biologie Évolutive UMR 5558, F-69622 Villeurbanne, France.\\
	$^{2}$École Normale Supérieure de Lyon, Université de Lyon, Université Lyon 1, Lyon, France\\
	$^{3}$Department of Biology, Institute of Biochemistry, and School of Mathematics and Statistics, Carleton University, Ottawa, Canada \\
	\texttt{\href{mailto:thibault.latrille@ens-lyon.org}{thibault.latrille@ens-lyon.org}} \\
}

\begin{document}
\maketitle

\begin{abstract}
	Adaptation in protein-coding sequences can be detected from multiple sequence alignments across species, or alternatively by leveraging polymorphism inside a population.
	Across species, quantification of the adaptive rate rely on phylogenetic codon models, classically formulated in terms of the ratio of non-synonymous over synonymous substitution rates, seeking signature of an accelerated rate of non-synonymous substitutions.
	However, because of the background of purifying selection, these models are potentially limited in their sensitivity.
	Recent developments have led to more sophisticated mutation-selection codon models aiming at making a more detailed quantitative assessment of the interplay between mutation, purifying and positive selection, leading to potentially more powerful and more quantitative models that have yet to be tested at large scale.
	In this study, we first conducted a large-scale exome analysis on placental mammals, and assessed the performance of mutation-selection codon models to detect proteins and sites under adaptation.
	Importantly, mutation-selection codon models are based on a population-genetic formalism and thus are directly comparable to McDonal \& Kreitman tests at the population level to quantify adaptation.
	Taking advantage of this relationship between phylogenetic and population genetics, we integrated divergence and polymorphism data across the entire exome for 29 populations across 7 genera, and showed that proteins and sites detected to be under adaptation at the phylogenetic scale are also under adaptation at the population-genetic scale.
	Altogether, our exome-wide analysis shows that phylogenetic mutation-selection codon models and population-genetic test of adaptation can be reconciled and are congruent, paving the way for integrative models and analyses across individuals and populations.
\end{abstract}

\keywords{Adaptation \and phylogenetic \and population genetics \and codon models}

\section*{Introduction}

Selection in molecular sequences takes a variety of forms, and a mutation can be either adaptive if positively selected, or instead purified if negatively selected.
In this context, present-day genetic sequences are informative of populations past evolutionary history and can carry signature of selection at different time scales.
One main goal of molecular evolution is to disentangle and quantify the intensity of neutral, adaptive and purifying evolution acting on sequences, leveraging variations in sequences between and within species.
Theoretically, in order to detect adaptive evolution, one must have data where part of the sequence is known to be under a neutral regime, which can be used as a null model.
In the case of protein-coding DNA sequences, synonymous sites are usually taken as proxies for neutral sites, although they may be under weak selection.
Non-synonymous mutations, on the other hand, might be under a mixture of adaptation and purifying selection.
Contrasting synonymous and non-synonymous changes, two different types of methods have emerged to quantify both positive and purifying selection acting on protein-coding sequences.
One method, stemming from phylogeny, uses multiple sequence alignment in different species to quantify adaptation.
Another method, stemming from population genetics, contrasts polymorphism inside a population and divergence to a close species.
One goal of the present work is to confront these two types of approaches.

In phylogeny-based method, the ratio of non-synonymous substitutions over synonymous substitutions, called $\omega$ is estimated from protein-coding DNA alignment\cite{muse_likelihood_1994,goldman_codon-based_1994}.
Assuming synonymous mutations are neutral, $\omega>1$ signals an excess in the rate of non-synonymous substitutions, indicating that the protein is under adaptive evolution.
Conversely, a default of non-synonymous substitutions, leading to $\omega<1$, means the protein is under purifying selection.
In practice, proteins are typically under a mix of adaptation and purifying selection, thus typically leading to an $\omega<1$ even in the presence of positive selection.
At a finer scale, site models detect specific site ($i$) of the sequence with a $\omega^{(i)}>1$\cite{yang_codon-substitution_2000, kosiol_patterns_2008}.
An alternative approach to detect adaptation would be to rely on an explicit \textit{nearly-neutral model} as the null model against which to detect deviation.
Recent development in this direction, the so-called phylogenetic mutation-selection models, provide a null model by estimating the fitness landscape over amino-acid sequences, for each site of the sequence\cite{yang_mutation-selection_2008, halpern_evolutionary_1998, rodrigue_mechanistic_2010}.
At the mutation-selection balance, the probability for a specific codon to be fixed in the population is proportional to its fitness, and mutation from a high fitness amino-acid towards a low fitness amino-acid will have a small probability of fixation, genuinely accounting purifying selection.
Conversely, only nearly-neutral mutations from high fitness amino acids will be permitted by the model, allowing to predict the nearly-neutral rate of non-synonymous substitutions at mutation-selection balance, called $\omega_{0}$.
By contrasting $\omega$ estimated by $\omega$-based codon models and $\omega_{0}$ predicted by mutation-selection model, one can hope to extract the rate of adaptation $\Delta \omega = \omega - \omega_{0}$.
However, mutation-selection models have not yet been applied to large-scale divergence data to quantify the rate of adaptation\cite{rodrigue_detecting_2016}.
In this study, we thus first applied $\omega$-based and mutation-selection codon models to whole exome on placental mammals, such as to quantify the rate $\Delta \omega$ for each site and protein and detect signature of adaptive evolution at the phylogenetic scale.

At the population-genetic scale, one of the most widely used tests for adaptation was proposed by McDonald and Kreitman\cite{mcdonald_adaptative_1991}.
This method uses the substitutions between two close species and polymorphism inside on population.
Deleterious mutations are quickly removed by selection, and the ratio of non-synonymous substitutions over synonymous substitutions ($\dnds$) is expected to be lower than one, since none of non-synonymous deleterious mutations will reach fixation.
Also the ratio of non-synonymous polymorphism over synonymous polymorphism ($\pnps$) is also expected to be lower than one, since the non-synonymous deleterious mutations will be removed quickly from the population.
Most importantly, in the absence of advantageous mutations, these two ratios are expected to be the same ($\dnds=\pnps$).
If advantageous mutations occur, then they are fixed rapidly in the population, thus contributing solely to divergence but not to polymorphism, leading to a strictly positive adaptive rate $\omega_{\mathrm{A}} = \dnds-\pnps$\cite{smith_adaptive_2002}.
This method is, however, plagued by the presence of moderately deleterious non-synonymous mutations, which can segregate at substantial frequency in the population without reaching fixation, thus contributing solely to polymorphism, and not to divergence, potentially resulting on an underestimation of the rate of adaptive evolution\cite{eyre-walker_quantifying_2002}.
Subsequent developments have tried to correct for this effect by relying on an explicit \textit{nearly-neutral model}, such as to derive an estimate of the rate of adaptation\cite{eyre-walker_estimating_2009, galtier_adaptive_2016}.

Altogether, the rate of adaptation is directly comparable between phylogenetic and population-genetic methods since they seek a deviation of $\omega$ from a nearly-neutral null model, estimated with mutation-selection models in phylogenetic context ($\omega_{0}$) or from standing polymorphism in a population-genetic context ($\pnps$).
Accordingly, the goal of this study is to assess whether the two signals of adaptation are correlated, which represent a unique opportunity to confront phylogenetic and population-genetic methods.
However, the population- and phylogeny-based method work over very different time scales, for that reason, they might be capturing different signals: long-term evolutionary Red-Queen for phylogeny based methods versus isolated events of adaptation along a particular lineage for population-based method.
Nonetheless, we expect sites and proteins under long-term evolutionary Red-Queen to maintain their signal of adaptation in several independent lineages.
To test this hypothesis, we develop a pipeline integrating divergence and polymorphism data across the entire exome for 29 populations across 7 genera, namely \textit{Equus}, \textit{Canis}, \textit{Bos}, \textit{Capra}, \textit{Ovis}, \textit{Chlorocebus} and \textit{Homo}.
The pipeline then testes if the group of sequences detected with a high rate of adaptation in the phylogeny-based method also display a high rate of adaptation in the population-based method, leveraging polymorphism available in each population.

\section*{Results}\label{sec:results}

\subsection*{Detecting genes and sites under adaptation}

We derived a two-step approach (see methods), applied to mammalian orthologs.
The value of $\omega$ estimated by the site model is shown as a function of $\omega_{0}$ estimated by the mutation-selection model for genes (scatter plot in figure~\ref{fig:scatterplot}A) and sites (density plot in figure~\ref{fig:scatterplot}B).
Genes and sites classified under an adaptive regime (in red) are rejecting the nearly-neutral assumption such that the lower bound for the confidence interval of $\omega$ is above the upper bound of the confidence interval of $\omega_{0}$, meaning that the value of their $\omega$ is too high compared to $\omega_{0}$.
Importantly, genes and sites classified under a adaptive regime have an average $\omega < 1$, otherwise they are considered as strongly adaptive, meaning that they could not have been by detected by homogeneous $\omega$-based codon models solely based on their value of $\omega$.
The nearly-neutral model assumption appears to be rejected for $564$ genes and $61.164$ sites, which are putatively under a long-term evolutionary Red-Queen.
On the other hand, the \textit{nearly-neutral} regime (in green) could not be rejected for $2868$ proteins and $1.707.826$ sites.

\subsection*{Ontology enrichment tests}

Next, we investigated whether the genes classified as adaptive showed enrichment in ontology terms.
Thus, we performed $490$ Fisher's exact test to estimates ontology enrichment by contrasting with the genes classified as under nearly-neutral regime.
$19$ ontologies are observed with a $p_{\mathrm{value}}$ corrected for multiple comparison (Holm–Bonferroni correction, $p_{\mathrm{value}}^{\mathrm{adjusted}}$) lower than the risk $\alpha=0.05$ (see table~\ref{table:ontology}).
The most enriched ontologies are immune system processes and innate immune response.
An enrichment in genes involved in immune processes was also found in large-scale analysis based on $\omega$-based site codon models\cite{kosiol_patterns_2008}.
Altogether, the mutation-selection method effectively detects adaptation regardless of the background of purifying selection, and returns reasonable candidates for adaptive evolution.

\subsection*{Congruence between phylogeny- and population-based methods}

Finally, we investigated whether the phylogeny-based and the population-based methods give congruent results in terms of detection of adaptive evolution.
At the population-genetic scale, $\omega_{\mathrm{A}}$ is the rate of adaptive evolution as proposed by McDonald and Kreitman\cite{mcdonald_adaptative_1991}.
On the one hand, the ratio of non-synonymous over synonymous polymorphism ($\pnps$) only contains non-adaptive polymorphisms.
On the other hand, divergence data with regard to a close specie allows one to estimate the ratio of non-synonymous over synonymous substitutions ($\dnds$) between these species, containing a mixture of both advantageous mutation and non-adaptive substitution.
As a result, $\omega_{\mathrm{A}} = \dnds - \pnps$ only contains the rate of adaptive substitution, and it should be higher for genes classified as adaptive in the phylogeny-based method, compared to genes classified as nearly-neutral.

For each $29$ population across $7$ genera and their sister species, $\omega_{\mathrm{A}}$ is computed on the concatenate of the $564$ genes classified as adaptive by the phylogeny-based method.
The result is compared to the empirical distribution of $\omega_{\mathrm{A}}$ over random sets of $564$ nearly-neutral genes (1000 sampling replicates), see figure~\ref{fig:method}.
Importantly, $\omega$ is controlled to be the same in the nearly-neutral replicate and the adaptive set of genes, such as to alleviate the fact that genes classified as adaptive have a higher $\omega$ than genes classified as nearly-neutral (see figure~\ref{fig:scatterplot}), which could bias our comparison since $\omega_{\mathrm{A}}$ could simply be higher for genes with higher $\omega$.
$p_{\mathrm{value}}$ are computed as the proportion of replicates for which $\omega_{\mathrm{A}}$ is higher in the nearly-neutral replicate than in the adaptive genes (computed only once).
The results are shown for the $29$ populations as violin plot in figure~\ref{fig:unfolded-MK}A and their value are shown in table~\ref{table:unfolded-MK}.
Out of $29$ populations, $24$ have a $p_{\mathrm{value}}^{\mathrm{adjusted}}$ lower than the risk $\alpha=0.05$ whenever corrected for multiple comparison (Holm–Bonferroni correction).
Of note, the only populations for which the test is not significant are human populations.

The same procedure is applied at a finer scale with sites instead of genes.
For each population, $\omega_{\mathrm{A}}$ is computed on the concatenate of the $61.164$ sites classified as adaptive by the phylogeny-based method, and compared to the empirical null distribution (figure~\ref{fig:unfolded-MK}B) and table~\ref{table:unfolded-MK}.
Out of $29$ populations, $8$ have a $p_{\mathrm{value}}^{\mathrm{adjusted}}$ lower than the risk $\alpha=0.05$ whenever corrected for multiple comparison (Holm–Bonferroni correction).
Of note, sites classified under an adaptive regime have an average $\omega < 1$, such that we discarded sites with $\omega > 1$ (considered under a strong adaptive regime), meaning we could detect long-term evolutionary Red-Queen even for a site with $\omega < 1$ that are still under adaptation at the population-genetic scale.

\section*{Discussion}\label{sec:discussion}

This study is the first large-scale application of the phylogenetic mutation-selection codon models for detecting adaptation.
This method contrast the observed ratio of non-synonymous over synonymous substitution rates ($\omega$) to a ratio predicted under a nearly-neutral regime ($\omega_{0}$)
Its application on mammalian genes suggests that $564$ out of $14477$ proteins are under long-term evolutionary Red-Queen, with ontology terms related to immune processes.
Mutation-selection models at the gene level has a somewhat a comparable sensitivity compared to site specific $\omega$-based codon models, for which $400$ genes were detected out of $16.529$ in a previous study~\cite{kosiol_patterns_2008}.
Moreover $61164$ sites are found to be under an adaptive regime ($\omega_{0} < \omega < 1$), which could theoretically never be detected by site-specific $\omega$-based codon models due to their $\omega$ lower than 1.

At the population-genetic scale, the availability of approaches to detect adaptation\cite{mcdonald_adaptative_1991, messer_frequent_2013} raises the question whether the signal of adaptation is congruent between phylogenetic and population genetics.
In this light, we aligned divergence and polymorphism data across the entire exome for 29 populations across 7 genera, namely \textit{Equus}, \textit{Canis}, \textit{Bos}, \textit{Capra}, \textit{Ovis}, \textit{Chlorocebus} and \textit{Homo}.
In $25$ out of $29$ populations, the set of genes detected to be under adaptation at the phylogenetic scale showed a significant increase in the rate of adaptation ($\omega_{\mathrm{A}} = \dnds - \pnps$) such as inferred by population-based method.
Notably, the $4$ populations for which the test is not conclusive are human populations, meaning that recent episode of adaptation in the \textit{Hominini} lineages does not reflect the long-term patterns of adaptation.
This result is in stark contrast with studies comparing $\omega$-based codon models at the gene level with MK methods, which found that the set of genes detected at different scales does not seem to overlap beyond random expectations\cite{chen_two_2021}.

Additionally, zooming in at the site level, for $8$ out of $29$ populations, the set of sites detected to be under adaptation at the phylogenetic scale showed a significant increase in $\omega_{\mathrm{A}}$ inferred by population-based method.
This result is congruent with other studies at the site level showing that $\omega$-based codon models are congruent with MK methods, with most adaptive mutations occurring at the surface of proteins in both methods\cite{moutinho_impact_2019}.
Moreover studies have shown that estimating the distribution of fitness effect at the population-genetic scale conditioning on SIFT scores estimated at the phylogenetic scale improves goodness of fit\cite{chen_hunting_2021}
In practice our result are in the sane direction, although our test might lack statistical power.
For example, there could be some background adaptation in proteins categorized here as being in a \textit{nearly-neutral} regime, which might not have been detected.
This explanation can be seen in the violin plots, since $\omega_{\mathrm{A}}$ is not equal to $0$ for genes and sites classified as nearly-neutral.
To confirm our results at the gene and site level that adaptation is congruent at different scales, our study could be replicated across different phylogeny, for examples in birds for which polymorphism datasets are readily available.

This study stands at the crossroad between the phylogenetic and population-genetic tests of adaptation and highlight their respective limitations.
On the one hand, phylogeny-based method is computationally intensive, requires a well-resolved species and gene trees, and it is crucially dependent on the quality of the alignments and paralogs identified as orthologs can easily falsify the results, hence the limitation to mammals in this study.
Additionally, all differences observed in the alignment are assumed to be substitutions, while some might in fact be polymorphisms segregating in the population.
Mistaking polymorphisms for substitutions is problematic, since both are not sensitive to mutation, selection and drift to the same extent\cite{mugal_why_2014}.
On the other hand, at the population-genetic scale, the greatest limitation is the lack of power determined by the genetic diversity since polymorphism are rare and estimation of $\pnps$ requires to pool many sites for which SNPs are available.
Furthermore, since the purifying process is stronger on longer time scales, $\omega_{\mathrm{A}}$ can be biased by moderately deleterious mutations\cite{eyre-walker_quantifying_2002, ho_time_2005} and by the change in population size through time\cite{eyre-walker_changing_2002}.
To overcome this bias, both Grapes\cite{galtier_adaptive_2016} and polyDFE\cite{tataru_polydfe_2020} are also used (supplementary materials), which relies on the synonymous and non-synonymous site-frequency spectra (SFS) to estimate the distribution of fitness effects of mutations (DFE), modeled as a continuous distribution.
However, the high range of $\omega_{\mathrm{A}}$ estimated on sets of genes/sites classified as nearly-neutral shows that these models lacking power (even more than the MK test) because of the sparsity of the SFS, stressing the limitations and difficulties of our study centered on confronting the rate of adaptive evolution at different time scales.

More broadly on a theoretical level, this study leverages only one specific overlap between phylogenetic and population genetics, namely that the rate of adaptation $\omega - \omega_{0}$ in phylogenetic codon models and $\dnds - \pnps$ in MK test should theoretically be directly comparable.
This comparison is grounded in the principle that the rate of non-adaptive substitution is approximated by $\pnps$ in the context of population genetics, which is referred to as $\omega_{0}$ in the context of the mutation-selection codon models.
Based on this theoretical relationship, our objective was to highlight that mutation-selection codon model provides a null (nearly-neutral) model from which we can disentangle purifying and adaptive evolution.
This positive result wields excitement to leverage mutation-selection models at a finer scale to quantify selection, rather than relying on a summary statistics (i.e. $\omega_{0}$).
As an example, the mutation-selection model provides a direct estimate of the selection coefficient for any mutation in a coding sequence, depending on the amino-acids involved in the mutation.
Such selection coefficients could be related to ontologies of SNPs, to assess whether mutation-selection codon models can predict the purifying selection process observed at the population-genetic scale.
In addition, aggregated over SNPs, the mutation-selection models provides an estimate of the distribution of fitness effect (DFE) and by extension would predict a site-frequency spectrum at the population-genetic scale, which could be compared to the empirical observation.
In this line of development, selection coefficients estimated by mutation-selection models could also be used instead of SIFT score to improve fit in polymorphism data\cite{chen_hunting_2021}.
These developments are readily available and can be tested using our generated datasets, leveraging the whole information available instead on focusing solely on the adaptive part of selection.
Altogether, our work is paving the way for studies and methods augmenting molecular polymorphism data within species with information about divergence data between species, and by assessing empirically the relationship between phylogenetic and population genetics\cite{thorne_codon_2012}.

\section*{Methods}\label{sec:methods}

\subsection*{Phylogenetic dataset}
protein-coding DNA sequences alignments in placental mammals are extracted from the \href{https://www.orthomam.univ-montp2.fr}{OrthoMaM} database\cite{ranwez_orthomam_2007, douzery_orthomam_2014, scornavacca_orthomam_2019}.
Genes located on the X, Y and mitochondrial chromosome are discarded from the analysis, since the number of polymorphism, necessary in population-based method, is expected to be different on these sequences.
Additionally, sequences from the species for which polymorphism are available, as well as their sister species have been discarded from the analysis to ensure independence between the data used in the phylogenetic and population-genetic method.

\subsection*{Adaptation in phylogeny-based method}
Classical codon models estimates a parameter $\omega=\dnds$, namely the ratio of the non-synonymous over the synonymous substitution rates\cite{muse_likelihood_1994,goldman_codon-based_1994}.
In the so-called site models, $\omega$ is allowed to vary across sites\cite{yang_codon-substitution_2000, huelsenbeck_dirichlet_2006}.
In \textit{Bayescode}, site-specific $\omega^{(i)}$ (figure~\ref{fig:scatterplot}B, y-axis) are independent identically distributed from a gamma distribution\cite{lartillot_phylobayes_2013}.
In a second step, the average over sites is calculated, giving estimates of $\omega$ for each protein-coding sequence (figure~\ref{fig:scatterplot}A, y-axis).

In contrast, mutation-selection models assume that the protein-coding sequence is at mutation-selection balance under a fixed fitness landscape, which is itself characterized by a fitness vector over the $20$ amino-acid at each site\cite{yang_mutation-selection_2008, halpern_evolutionary_1998, rodrigue_mechanistic_2010}.
Mathematically, the rate of non-synonymous substitution from codon $a$ to codon $b$ ($q_{a \mapsto b}^{(i)}$) at site $i$ of the sequence is equal to the rate of mutation from the underlying DNA change ($\mu_{a \mapsto b}$) multiplied by the scaled probability of fixation of the mutation ($\proba_{a \mapsto b}^{(i)}$).
Crucially, the probability of fixation depends on the difference of scaled fitness between the amino-acid encoded by the mutated codon ($F_b^{(i)}$) and the fitness of the amino-acid encoded by the original codon ($F_a^{(i)}$) of site $i$\cite{wright_evolution_1931, fisher_genetical_1930}.
Altogether, the rate of substitution from codon $a$ to $b$ at a given site $i$ is:
\begin{equation}
	q_{a \mapsto b}^{(i)} = \mu_{a \mapsto b} \dfrac{F_b^{(i)} - F_a^{(i)}}{1 - \e^{F_a^{(i)} - F_b^{(i)}}}.
\end{equation}

Fitting the mutation-selection model on a sequence alignment leads to an estimation of the mutation rate matrix ($\UniDimArray{mu}$) as well as the 20 amino-acid fitness landscape ($\UniDimArray{f^{(i)}}$) at each site $i$.
From these parameters, one can compute $\omega_{0}^{(i)}$ (figure~\ref{fig:scatterplot}B, x-axis), the site-specific predicted rate of non-synonymous over synonymous substitution at the mutation-selection balance:
\begin{equation}
	\omega_{0}^{(i)} = \dfrac{\sum_{a \in \mathcal{C}} \sum_{b \in \mathcal{N}_a} \pi_a^{(i)} q_{a \mapsto b}^{(i)}}{\sum_{a \in \mathcal{C}} \sum_{b \in \mathcal{N}_a} \pi_a^{(i)} \mu_{a \mapsto b}},
\end{equation}
where $\mathcal{C}$ is the set all the possible codons ($61$ by discarding stop codons), $\pi_a^{(i)}$ is the equilibrium frequency of codon $a$ at site $i$, and $\mathcal{N}_a$ is the set of codons that are non-synonymous to $a$\cite{spielman_relationship_2015, rodrigue_detecting_2016}.
In a second step, the average over sites is calculated, giving estimates of $\omega_{0}$ for each protein-coding sequences (figure~\ref{fig:scatterplot}A, x-axis).
Under the assumption that the protein is under a \textit{nearly-neutral} regime, the predicted $\omega_{0}$ (mutation-selection model) and the estimated $\omega$ (site model) should be the same\cite{spielman_relationship_2015}.

We ran the Bayesian software \href{https://github.com/bayesiancook/bayescode}{BayesCode} on each protein-coding DNA alignment\cite{lartillot_phylobayes_2013, rodrigue_detecting_2016}.
Each Monte-Carlo Markov-Chain (MCMC) is run during $2000$ points, with a burn-in of $1000$ points, allowing to compute the mean of $\omega$ and $\omega_{0}$ across the MCMC, as well as the $99.95$\% confidence interval for genes $90$\% confidence interval for sites.
Genes and sites classified under a strongly adaptive regime (in black) have an average $\omega > 1$.
Genes and sites classified under an adaptive regime (in red) are rejecting the nearly-neutral assumption such that a lower bound for the confidence interval of $\omega$ is above the upper bound of the confidence interval of $\omega_{0}$, meaning $\omega > \omega_{0}$.
Genes and sites classified under an epistasis regime (in blue) are such that a higher bound for the confidence interval of $\omega$ is below the lower bound of the confidence interval of $\omega_{0}$, meaning $\omega < \omega_{0}$.
Genes and sites are classified under a nearly-neutral regime (in green) if the average $\omega$ is within the confidence interval of the $\omega_{0}$, and respectively the average $\omega_{0}$ is also within the confidence interval of  $\omega$, meaning $\omega = \omega_{0}$.
Genes and sites that do not fall in any of these categories are considered unclassified.

\subsection*{Polymorphism dataset}

In order to compare polymorphism and divergence, each SNP (chromosome, position, strand) in the focal species must be matched to its relative position in the protein-coding DNA alignment.
First, genomic positions are converted to relative position in the coding sequence (CDS) using gene annotation files (GTF format) downloaded from Ensembl (\url{ensembl.org}), we verified the SNP match the reference in the CDS (FASTA format) also downloaded from Ensembl.
Secondly, relative position in the CDS is converted to position in the multiple sequence alignment (containing gaps) from OrthoMaM database\cite{ranwez_orthomam_2007, douzery_orthomam_2014, scornavacca_orthomam_2019} by doing a global pairwise alignment (Biopython pairwise2) between the CDS fasta and the sequence found in the alignment.
This conversion from genomic position to position in the alignment is only possible if the assembly used for SNP calling is the same as the one used in the alignment, the GTF annotations and the FASTA sequences.
We retrieved polymorphism from different projects.
\textit{Equus caballus} variants called on the EquCab2 assembly in the EVA study PRJEB9799, we considered \textit{Ceratotherium simum simum} as its sister species.
\textit{Canis familiars} variants called on the CanFam3.1 assembly in the EVA study PRJEB24066, we considered \textit{Ursus maritimus} as its sister species.
\textit{Bos taurus} variants called on the UMD3.1 assembly in the next-gen project, we considered \textit{Bison bison bison} as its sister species.
\textit{Ovis aries} variants called on the Oar\_v3.1 assembly in the next-gen project, we considered \textit{Pantholops hodgsonii} as its sister species.
\textit{Capra Hircus} variants called on the CHIR1 assembly in the next-gen project, we performed a liftover to the ARS1 assembly and considered \textit{Pantholops hodgsonii} as its sister species.
\textit{Chlorocebus sabaeus} variants are called on the ChlSab1.1 assembly in the EVA study PRJEB22989\cite{svardal_ancient_2017}, we considered \textit{Macaca mulatta} as its sister species.
\textit{Homo sapiens} variants are called on the GRCh38 assembly in the 1000-genome project\cite{consortium_integrated_2012, the_1000_genomes_project_consortium_global_2015}, we considered \textit{Pan troglodytes} as its sister species.
Variants not inside genes are discarded at the beginning of the analysis.
Insertions and deletions are not analyzed, and only Single Nucleotide Polymorphisms (SNPs) with only one mutant allele are considered.
Stop codon mutants are also discarded.
For populations containing more than $8$ sampled individuals, the site-frequency spectrum (SFS) is subsampled down to $16$ chromosomes ($8$ diploid individuals) without replacement (hyper-geometric distribution) to alleviate the effect of different sampling depth in the $29$ populations.
Moreover, subsampling mitigate the impact of moderately deleterious mutations segregating at low frequency on $\pnps$, since they are more likely to be discarded than polymorphism segregating at higher frequency.
The Snakemake pipeline for integrating polymorphism and divergence data uses custom scripts written in python 3.8.

\subsection*{Rate of adaption in population-based method}

The genes and sites classified as under adaptation are concatenated.
For each population $\pnps$ is computed as the sum of non-synonymous over synonymous polymorphism on the concatenated SFS\@.
$\dnds$ is computed on the concatenated pairwise alignment between focal and sister species extracted from OrthoMaM, the $\dnds$ count is performed by \textit{yn00}.
Altogether, $\omega_{\mathrm{A}} = \dnds - \pnps$ is thus computed for each population on genes and sites classified as under adaptation.

The result is compared to the empirical null distribution of $\omega_{\mathrm{A}}$, obtained by randomly sampling (1000 sampling replicates) a subset of genes/sites classified as nearly-neutral.
The random sampling is weighted to control for $\omega$ in the set of nearly-neutral genes/sites.
First, a normal distribution is fitted to $\omega$ in both sets, and the probability density is called $f$ for the adaptive set and $g$ for the nearly-neutral set.
Secondly, for each gene/site classified as nearly-neutral the weight is computed as the ratio $f(\omega)/g(\omega)$ for this specific gene/site.
Sampling with this procedure produce a set of genes/sites classified as nearly-neutral with the same $\omega$ on average than the set of adaptive genes/sites.

Other methods to compute $\omega_{\mathrm{A}}$ such as Grapes\cite{galtier_adaptive_2016} and polyDFE\cite{tataru_polydfe_2020} are also used (supplementary materials), which relies on the synonymous and non-synonymous unfolded site-frequency spectra (SFS) to estimate the distribution of fitness effects of mutations (DFE), modeled as a continuous distribution.
In Grapes, GammaExpo models the fitness effect of weakly deleterious non-synonymous mutations as distributed according to a negative Gamma and the fitness effect of weakly advantageous mutations are distributed exponentially.
This method is an extension of the methods introduced by Eyre-Walker and collaborators\cite{eyre-walker_distribution_2006, eyre-walker_estimating_2009}.
Unfolded SFSs are obtained by polarizing SNPs using the $3$ closest outgroups found in the OrthoMam alignment with est-usfs\cite{keightley_inferring_2018}.

\section{Data availability}\label{sec:data-availability}

The data underlying this article are available in Github, at \url{https://github.com/ThibaultLatrille/AdaptaPop}, as well as scripts and instructions necessary to reproduce the empirical experiments on the original dataset or with user-specified datasets.

\section{Acknowledgements}\label{sec:acknowledgements}

We gratefully also acknowledge the help of Nicolas Galtier for his advice and review concerning this manuscript.
This work was performed using the computing facilities of the CC LBBE/PRABI.
Funding: French National Research Agency, Grant ANR-15-CE12-0010-01 / DASIRE.

\section{Author information}\label{sec:author-information}

TL gathered and formatted the data, developed the new models in \texttt{BayesCode} and conducted the analyses.
TL, NR and NL contributed to the writing of the manuscript.

\printbibliography

\begin{figure*}[!ht]
	\centering
	\begin{minipage}{0.49\linewidth}
		\includegraphics[width=\linewidth, page=1]{scatterplot_gene}
	\end{minipage}
	\llap{\raisebox{1.8cm}{\scriptsize A\hspace{7.0cm}}}\hfill
	\begin{minipage}{0.49\linewidth}
		\includegraphics[width=\linewidth, page=1]{scatterplot_site}
	\end{minipage}
	\llap{\raisebox{1.8cm}{\scriptsize B\hspace{7.0cm}}}\hfill
	\caption{
		Detection of protein-coding sequences ongoing adaptation at the phylogenetic scale.
		$\omega$ estimated by the site model against $\omega_{0}$ predicted by the mutation-selection model.
		Scatter plot of $14475$ genes in panel A, with $99.95$\% confidence interval. Density plot of $2636948$ sites in panel B.
		Genes or sites are then classified into one of the four evolutionary regimes: strongly adaptive ($\omega > 1$ in black), adaptive ($1 > \omega > \omega_{0}$ in red), nearly-neutral ($\omega \simeq \omega_{0}$ in green) or epistasis ($\omega < \omega_{0}$ in blue).}
	\label{fig:scatterplot}
\end{figure*}

\begin{table*}[!ht]
	\centering
	\begin{adjustbox}{width = 1\textwidth}
		\small
		\begin{tabular}{|l|r|r|r|r|r|}
			\toprule
			\textbf{Ontology} &
			$\bm{n_{\mathrm{Observed}}}$ &
			$\bm{n_{\mathrm{Expected}}}$ &
			\textbf{Odds ratio} &
			$\bm{p_{\mathrm{value}}}$ &
			$\bm{p_{\mathrm{value}}^{\mathrm{adjusted}}}$ \\
			\midrule
			immune system process & 38 & 4.868 & 7.806 & 1.6$\times 10^{-14}$ & 7.9$\times 10^{-12}$$\bm{^*}$ \\
			innate immune response & 32 & 4.253 & 7.524 & 3$\times 10^{-12}$ & 1.5$\times 10^{-9}$$\bm{^*}$ \\
			extracellular space & 60 & 20.3 & 2.962 & 4.3$\times 10^{-9}$ & 2.1$\times 10^{-6}$$\bm{^*}$ \\
			extracellular region & 73 & 29.3 & 2.494 & 4.3$\times 10^{-8}$ & 2.1$\times 10^{-5}$$\bm{^*}$ \\
			cell surface & 29 & 6.634 & 4.371 & 8$\times 10^{-8}$ & 3.9$\times 10^{-5}$$\bm{^*}$ \\
			external side of plasma membrane & 17 & 2.354 & 7.221 & 4.7$\times 10^{-7}$ & 0.00023$\bm{^*}$ \\
			blood microparticle & 14 & 1.385 & 10.1 & 5.5$\times 10^{-7}$ & 0.00027$\bm{^*}$ \\
			regulation of complement activation & 10 & 0.401 & 24.9 & 9.6$\times 10^{-7}$ & 0.00046$\bm{^*}$ \\
			defense response to virus & 12 & 0.997 & 12.0 & 1.5$\times 10^{-6}$ & 0.00073$\bm{^*}$ \\
			plasma membrane & 98 & 48.8 & 2.009 & 2.2$\times 10^{-6}$ & 0.001$\bm{^*}$ \\
			immune response & 22 & 5.447 & 4.039 & 6.2$\times 10^{-6}$ & 0.003$\bm{^*}$ \\
			platelet degranulation & 11 & 1.001 & 11.0 & 6.5$\times 10^{-6}$ & 0.003$\bm{^*}$ \\
			integral component of plasma membrane & 46 & 19.5 & 2.355 & 1.6$\times 10^{-5}$ & 0.008$\bm{^*}$ \\
			chemotaxis & 11 & 1.404 & 7.837 & 3.4$\times 10^{-5}$ & 0.016$\bm{^*}$ \\
			proteolysis & 25 & 7.959 & 3.141 & 3.5$\times 10^{-5}$ & 0.016$\bm{^*}$ \\
			receptor-mediated endocytosis & 10 & 1.207 & 8.283 & 6$\times 10^{-5}$ & 0.029$\bm{^*}$ \\
			positive regulation of ERK1 and ERK2 cascade & 13 & 2.396 & 5.426 & 6.6$\times 10^{-5}$ & 0.031$\bm{^*}$ \\
			cell surface receptor signaling pathway & 17 & 4.354 & 3.905 & 8.8$\times 10^{-5}$ & 0.041$\bm{^*}$ \\
			extracellular exosome & 66 & 34.5 & 1.912 & 8.9$\times 10^{-5}$ & 0.042$\bm{^*}$ \\
			adaptive immune response & 12 & 2.204 & 5.445 & 0.00012 & 0.058\hspace{5pt} \\
			serine-type endopeptidase activity & 14 & 3.192 & 4.386 & 0.00016 & 0.073\hspace{5pt} \\
			inflammatory response & 20 & 6.304 & 3.173 & 0.00017 & 0.081\hspace{5pt} \\
			\bottomrule
		\end{tabular}
	\end{adjustbox}
	\caption{
		Ontology enrichment in the adaptive genes.
		490 ontology Fisher's exact test are performed with $243$ genes detected as under adaptation and $1164$ as under \textit{nearly-neutral} regime for which ontologies are available.
		$p_{\mathrm{value}}^{\mathrm{adjusted}}$ are corrected for multiple comparison (Holm–Bonferroni correction).
		$^*$ for $p_{\mathrm{value}}^{\mathrm{adjusted}} < 0.05$.
	}
	\label{table:ontology}
\end{table*}

\begin{figure*}[!ht]
	\centering
	\includegraphics[width=\linewidth, page=1]{polymorphism-method}
	\caption{
		Integrating divergence and polymorphism for the detection of adaptation.
		At the phylogenetic level, $\omega$ (classical codon models) and $\omega_{0}$ (mutation-selection codon models) are computed from protein-coding DNA alignments, allowing to classify genes into adaptive (in red) and nearly-neutral (in green) regime.
		At the population-genetic level, for each population, $\omega_{\mathrm{A}}$ is computed on the concatenate of genes classified as under adaptation.
		The result is compared to the empirical null distribution of $\omega_{\mathrm{A}}$ in each population, obtained by randomly sampling (1000 replicates) a subset under a nearly-neutral regime, for which $\omega$ is controlled to be the same in both sets of genes.
	}
	\label{fig:method}
\end{figure*}

\begin{figure*}[!ht]
	\centering
	\includegraphics[width=\linewidth, page=1]{unfolded-MK}
	\caption{
		Enrichment of adaptation at the population-genetic scale for $29$ populations across $7$ genera at the gene (panel A) and site (panel B) level.
		For each population, $\omega_{\mathrm{A}}$ is computed on $564$ genes and $61164$ sites having a high rate of adaptation at the phylogenetic scale (in red).
		The result is compared to the empirical null distribution of $\omega_{\mathrm{A}}$, obtained by randomly sampling (1000 replicates) a subset of genes and sites under a nearly-neutral regime (violin plot in green).
		$^*$ signify that the $p_{\mathrm{value}}$ corrected for multiple comparison (Holm–Bonferroni correction) is lower than the risk $\alpha=0.05$.
		The acronym of populations, and the quantitative value of $\omega_{\mathrm{A}}$ and $p_{\mathrm{value}}$ are shown in table~\ref{table:unfolded-MK}}
	\label{fig:unfolded-MK}
\end{figure*}

\begin{table*}[!ht]
	\centering
	\begin{adjustbox}{width = 1\textwidth}
		\begin{tabular}{|l|l|r|r|r|r|r|r|r|r|r|}
			\toprule
			\multicolumn{2}{|c|}{} & \multicolumn{4}{c|}{Genes} & \multicolumn{4}{c|}{Sites} \\ \hline
			\textbf{Species} &
			\textbf{Population} &
			\specialcell{$\bm{\omega_{\mathrm{A}}}$ \\ adaptive} &
			\specialcell{$\bm{\left< \omega_{\mathrm{A}} \right>}$ \\ nearly-neutral} &
			$\bm{p_{\mathrm{value}}}$ &
			$\bm{p_{\mathrm{value}}^{\mathrm{adjusted}}}$ &
			\specialcell{$\bm{\omega_{\mathrm{A}}}$ \\ adaptive}  &
			\specialcell{$\bm{\left< \omega_{\mathrm{A}} \right>}$ \\ nearly-neutral} &
			$\bm{p_{\mathrm{value}}}$ &
			$\bm{p_{\mathrm{value}}^{\mathrm{adjusted}}}$ &
			$\bm{\pi_{\textrm{S}}}$ \\
			\midrule
			Equus caballus & Diverse (Equus) & 0.102 & -0.036 & 0.0 & 0.0$\bm{^*}$ & 0.150 & 0.015 & 0.040 & 0.440\hspace{5pt} & 0.002 \\ \hline
			\rowcolor{LIGHTGREY} Canis familiaris & Diverse (Canis) & 0.091 & -0.014 & 0.0 & 0.0$\bm{^*}$ & 0.169 & 0.010 & 0.002 & 0.044$\bm{^*}$ & 0.003 \\ \hline
			Bos taurus & Iran (IRBT) & 0.143 & 0.109 & 0.005 & 0.036$\bm{^*}$ & 0.183 & 0.065 & 0.005 & 0.095\hspace{5pt} & 0.008 \\
			Bos taurus & Uganda (UGBT) & 0.148 & 0.112 & 0.004 & 0.036$\bm{^*}$ & 0.204 & 0.061 & 0.0 & 0.0$\bm{^*}$ & 0.008 \\ \hline
			\rowcolor{LIGHTGREY} Capra hircus & Australia (AUCH) & 0.096 & 0.032 & 0.0 & 0.0$\bm{^*}$ & 0.118 & 0.013 & 0.073 & 0.584\hspace{5pt} & 0.003 \\
			\rowcolor{LIGHTGREY} Capra hircus & France (FRCH) & 0.091 & 0.016 & 0.0 & 0.0$\bm{^*}$ & 0.159 & 0.053 & 0.060 & 0.540\hspace{5pt} & 0.002 \\
			\rowcolor{LIGHTGREY} Capra aegagrus & Iran (IRCA) & 0.120 & 0.041 & 0.0 & 0.0$\bm{^*}$ & 0.120 & 0.052 & 0.125 & 0.875\hspace{5pt} & 0.003 \\
			\rowcolor{LIGHTGREY} Capra hircus & Iran (IRCH) & 0.106 & 0.030 & 0.0 & 0.0$\bm{^*}$ & 0.150 & 0.025 & 0.018 & 0.288\hspace{5pt} & 0.004 \\
			\rowcolor{LIGHTGREY} Capra hircus & Italy (ITCH) & 0.090 & 0.012 & 0.0 & 0.0$\bm{^*}$ & 0.185 & 0.061 & 0.020 & 0.300\hspace{5pt} & 0.003 \\
			\rowcolor{LIGHTGREY} Capra hircus & Morocco (MOCH) & 0.107 & 0.026 & 0.0 & 0.0$\bm{^*}$ & 0.108 & 0.011 & 0.049 & 0.490\hspace{5pt} & 0.004 \\ \hline
			Ovis aries & Iran (IROA) & 0.173 & 0.078 & 0.0 & 0.0$\bm{^*}$ & 0.202 & 0.043 & 0.0 & 0.0$\bm{^*}$ & 0.007 \\
			Ovis orientalis & Iran (IROO) & 0.180 & 0.079 & 0.0 & 0.0$\bm{^*}$ & 0.211 & 0.058 & 0.0 & 0.0$\bm{^*}$ & 0.008 \\
			Ovis vignei & Iran (IROV) & 0.174 & 0.105 & 0.0 & 0.0$\bm{^*}$ & 0.254 & 0.073 & 0.0 & 0.0$\bm{^*}$ & 0.005 \\
			Ovis aries & Various (ISGC) & 0.171 & 0.079 & 0.0 & 0.0$\bm{^*}$ & 0.192 & 0.052 & 0.0 & 0.0$\bm{^*}$ & 0.008 \\
			Ovis aries & Morocco (MOOA) & 0.176 & 0.081 & 0.0 & 0.0$\bm{^*}$ & 0.217 & 0.053 & 0.0 & 0.0$\bm{^*}$ & 0.007 \\ \hline
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Barbados & 0.107 & 0.018 & 0.0 & 0.0$\bm{^*}$ & 0.214 & -0.041 & 0.0 & 0.0$\bm{^*}$ & 0.003 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Central African Republic (CAR) & 0.101 & 0.045 & 0.0 & 0.0$\bm{^*}$ & 0.100 & -0.00019 & 0.027 & 0.351\hspace{5pt} & 0.006 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Ethiopia & 0.055 & 0.007 & 0.005 & 0.036$\bm{^*}$ & 0.158 & 0.018 & 0.004 & 0.080\hspace{5pt} & 0.005 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Gambia & 0.083 & 0.029 & 0.0 & 0.0$\bm{^*}$ & 0.150 & -0.004 & 0.003 & 0.063\hspace{5pt} & 0.004 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Kenya & 0.101 & 0.046 & 0.0 & 0.0$\bm{^*}$ & 0.106 & -0.020 & 0.021 & 0.300\hspace{5pt} & 0.004 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Nevis & 0.058 & 0.006 & 0.003 & 0.030$\bm{^*}$ & 0.145 & -0.025 & 0.006 & 0.108\hspace{5pt} & 0.003 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & South Africa (SA) & 0.095 & 0.025 & 0.0 & 0.0$\bm{^*}$ & 0.100 & 0.00034 & 0.028 & 0.351\hspace{5pt} & 0.006 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Saint Kitts (SK) & 0.064 & 0.010 & 0.005 & 0.036$\bm{^*}$ & 0.111 & -0.034 & 0.009 & 0.153\hspace{5pt} & 0.004 \\
			\rowcolor{LIGHTGREY} Chlorocebus sabaeus & Zambia & 0.095 & 0.019 & 0.0 & 0.0$\bm{^*}$ & 0.023 & -0.008 & 0.291 & 1.000\hspace{5pt} & 0.006 \\ \hline
			Homo sapiens & African (AFR) & 0.057 & 0.009 & 0.037 & 0.111\hspace{5pt} & -0.069 & 0.009 & 0.810 & 1.000\hspace{5pt} & 0.002 \\
			Homo sapiens & Ad Mixed American (AMR) & 0.060 & -0.014 & 0.004 & 0.036$\bm{^*}$ & -0.035 & -0.004 & 0.660 & 1.000\hspace{5pt} & 0.002 \\
			Homo sapiens & East Asian (EAS) & 0.011 & -0.010 & 0.248 & 0.248\hspace{5pt} & -0.011 & -0.013 & 0.506 & 1.000\hspace{5pt} & 0.002 \\
			Homo sapiens & European (EUR) & 0.029 & -0.019 & 0.046 & 0.111\hspace{5pt} & -0.079 & -0.002 & 0.804 & 1.000\hspace{5pt} & 0.002 \\
			Homo sapiens & South Asian (SAS) & 0.038 & -0.024 & 0.014 & 0.056\hspace{5pt} & -0.199 & -0.007 & 0.963 & 1.000\hspace{5pt} & 0.002 \\
			\bottomrule
		\end{tabular}
	\end{adjustbox}
	\caption{
		Table of quantitative value of $\omega_{\mathrm{A}}$ across $29$ population shown in figure~\ref{fig:unfolded-MK} and the $p_{\mathrm{value}}$ associated to the comparison between the genes and sites classified as adaptive and nearly-neutral.
		$\left< \right>$ refers to the mean over the $1000$ replicates (subsampling).
		$\pi_{\textrm{S}}$ is the observed genetic diversity (number of SNPs per site) counted over synonymous sites.
		$p_{\mathrm{value}}^{\mathrm{adjusted}}$ are corrected for multiple comparison (Holm–Bonferroni correction).
		$^*$ for $p_{\mathrm{value}}^{\mathrm{adjusted}} < 0.05$.
	}
	\label{table:unfolded-MK}
\end{table*}

\end{document}

