%! BibTeX Compiler = biber
\documentclass{article}
\usepackage{caption}
\usepackage{xcolor}
\definecolor{BLUELINK}{HTML}{0645AD}
\definecolor{DARKBLUELINK}{HTML}{0B0080}
\definecolor{LIGHTBLUELINK}{HTML}{3366BB}
\definecolor{PURPLELINK}{HTML}{663366}
\PassOptionsToPackage{hyphens}{url}
\usepackage[colorlinks=false ]{hyperref}
% for linking between references, figures, TOC, etc in the pdf document
\hypersetup{colorlinks,
    linkcolor=DARKBLUELINK,
    anchorcolor=DARKBLUELINK,
    citecolor=DARKBLUELINK,
    filecolor=DARKBLUELINK,
    menucolor=DARKBLUELINK,
    urlcolor=BLUELINK
} % Color citation links in purple
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{naturalnames}{hyperref}

\usepackage[margin=60pt]{geometry}
\usepackage{amssymb,amsfonts,amsmath,amsthm,mathtools}
\usepackage{lmodern}
\usepackage{bm,bbold}
\usepackage{bbm}
\usepackage{verbatim}
\usepackage{float}
\usepackage{listings, enumerate, enumitem}
\usepackage[export]{adjustbox}
\usepackage{tabu}
\usepackage{longtable}
\tabulinesep=0.6mm
\newcommand\cellwidth{\TX@col@width}
\usepackage{hhline}
\setlength{\arrayrulewidth}{1.2pt}
\usepackage{multicol,multirow,array}
\usepackage{etoolbox}
\AtBeginEnvironment{tabu}{\footnotesize}
\usepackage{booktabs}

\usepackage{graphicx}
\graphicspath{{artworks/}}
\makeatletter
\def\input@path{{artworks/}}
\makeatother
\pdfstringdefDisableCommands{%
    \renewcommand*{\bm}[1]{#1}%
% any other necessary redefinitions
}
\newcommand{\specialcell}[2][c]{%
    \begin{tabular}[#1]{@{}c@{}}
        #2
    \end{tabular}}
\renewcommand{\thetable}{S\arabic{table}}
\renewcommand{\thefigure}{S\arabic{figure}}

\usepackage{xfrac, nicefrac}
\usepackage[backend=biber,isbn=false,url=false,intitle=true,style=nature]{biblatex}
\addbibresource{codon_models.bib}
\pdfinclusioncopyfonts=1

\newcommand{\UniDimArray}[1]{\bm{#1}}
\newcommand{\BiDimArray}[1]{\bm{#1}}
\DeclareMathOperator{\E}{\mathbb{E}}
\DeclareMathOperator{\Var}{\mathrm{Var}}
\newcommand{\der}{\mathrm{d}}
\newcommand{\angstrom}{\mathrm{\normalfont\AA}}
\newcommand{\e}{\mathrm{e}}
\newcommand{\avg}[1]{\left< #1 \right>} % for average
\newcommand{\Ne}{N_{\mathrm{e}}}
\newcommand{\dn}{d_N}
\newcommand{\ds}{d_S}
\newcommand{\dnds}{\dn / \ds}
\newcommand{\rateApop}{\omega_{\mathrm{A}}}
\newcommand{\rateAphy}{\rateApop^{\mathrm{phy}}}
\newcommand{\pn}{\pi_N}
\newcommand{\ps}{\pi_S}
\newcommand{\pnps}{\pn / \ps}
\newcommand{\proba}{\mathbb{P}}
\newcommand{\pfix}{\proba_{\mathrm{fix}}}
\newcommand{\Pfix}{2 \Ne \proba_{\mathrm{fix}}}
\newcommand{\indice}{a}
\newcommand{\indiceexp}{^{(\indice)}}

\newcommand{\Sphy}{S}
\newcommand{\Spop}{\beta}

\title{Genes and sites under adaptation at the phylogenetic scale also exhibit adaptation at the population-genetic scale}

%\date{September 9, 1985}	% Here you can change the date presented in the paper title
%\date{} \rowcolor{LIGHTGREY}% Or removing it

\author{
    \large
    \textbf{T. {Latrille}$^{1,2,3}$, N. {Rodrigue}$^{4}$, N. {Lartillot}$^{1}$}\\
    \footnotesize $^{1}$Université de Lyon, CNRS, LBBE UMR 5558, Villeurbanne, France\\
    \footnotesize $^{2}$École Normale Supérieure de Lyon, Université de Lyon, Lyon, France\\
    \footnotesize $^{3}$Université de Lausanne, Lausanne, Switzerland\\
    \footnotesize $^{4}$Department of Biology, Carleton University, Ottawa, Canada \\
    \footnotesize \texttt{\href{mailto:thibault.latrille@ens-lyon.org}{thibault.latrille@ens-lyon.org}} \\
}

\date{}

\begin{document}
    \maketitle
    \part*{Supplementary materials}
    \tableofcontents
    \pagebreak


    \section{Running Bayescode}

    \subsection{Site-specific $\omega$-based codon models.}
    The $61$-by-$61$ codon substitution matrix ($\UniDimArray{q^{(i)}}$) at site $i$ is defined entirely by the mutation matrix ($\UniDimArray{\mu}$), $\omega^{(i)}$ and the genetic code:
    \begin{equation}
        \begin{dcases}
            q_{a \mapsto b}^{(i)} & = 0 \text{ if codons $a$ and $b$ are more than one mutation away,} \\
            q_{a \mapsto b}^{(i)} & = \mu_{a \mapsto b} \text{ if codons $a$ and $b$ are synonymous,} \\
            q_{a \mapsto b}^{(i)} & = \omega^{(i)} \mu_{a \mapsto b} \text{ if codons $a$ and $b$ are non-synonymous}.
        \end{dcases}
    \end{equation}
    By definition of the instantaneous rate matrix, the sum of the entries in each row of the codon substitution rate matrix $\UniDimArray{q}$ is equal to $0$, giving the diagonal entries:
    \begin{equation}
        q_{a \mapsto b}^{(i)}= - \sum\limits_{b \neq a, b = 1}^{61} q_{a \mapsto b}^{(i)}.
    \end{equation}

    In BayesCode (\url{https://github.com/ThibaultLatrille/bayescode}), $\omega$-based site-specific codon models are obtained by running \textit{mutselomega} with the options:
    \begin{scriptsize}
        \begin{verbatim}
        mutselomega --omegashift 0.0 --freeomega --omegancat 30 --flatfitness -a my_alignment.phy -t my_tree.newick -u 2000 my_genename
        \end{verbatim}
    \end{scriptsize}
    The mean value of $\omega$ per site is then obtained by running \textit{readmutselomega} with the options:
    \begin{scriptsize}
        \begin{verbatim}
        readmutselomega --every 1 --until 2000 --burnin 1000 -c 0.025 my_genename
        \end{verbatim}
    \end{scriptsize}

    \subsection{Site-specific mutation-selection codon models}

    In BayesCode (\url{https://github.com/ThibaultLatrille/bayescode}), mutation-selection codon models are obtained by running \textit{mutselomega} for 2000 points of MCMC with the options:
    \begin{scriptsize}
        \begin{verbatim}
        mutselomega ---omegashift 0.0 --ncat 30 -a my_alignment.phy -t my_tree.newick -u 2000 my_genename
        \end{verbatim}
    \end{scriptsize}
    The collection of site-specific fitness profiles ($\UniDimArray{F^{(i)}}, \forall i$) are then obtained by running \textit{readmutselomega}, reading 1000 points of MCMC (first 1000 are considered as burn-in) with the options:
    \begin{scriptsize}
        \begin{verbatim}
        readmutselomega --every 1 --until 2000 --burnin 1000 --ss my_genename
        \end{verbatim}
    \end{scriptsize}
    The gene-specific mutation matrix ($\UniDimArray{\mu}$) is also obtained by running \textit{readmutselomega}, reading 1000 points of MCMC (first 1000 are considered as burn-in) with the options:
    \begin{scriptsize}
        \begin{verbatim}
        readmutselomega --every 1 --until 2000 --burnin 1000 --nuc my_genename
        \end{verbatim}
    \end{scriptsize}

    \newpage


    \section{Gene ontology enrichment at gene and site level}
    \label{sec:gene-ontology-enrichment}

    \subsection{Gene-specific mutation-selection model}

    Genes are classified under an adaptive regime if the lower bound for the posterior credibility interval of gene-specific $\omega$ ($\alpha=0.05$) is above the upper bound of the posterior credibility interval of gene-specific $\omega_{0}$.
    Because this is a unilateral test ($\omega > \omega_{0}$) and the two posterior credibility interval are independent, the risk is $(\alpha/2)^2=0.025^2=6.25 \times 10^{-4}$ for each gene.
    Genes are classified as control if they are not in the adaptive group.
    For each ontology ($775$ ontologies), a 2x2 contingency tables is built by counting the number of genes based on their evolutionary regime (adaptive regime or control group) and their ontology (whether they have this specific ontology or not).
    Fisher's exact tests are then performed for these 2x2 contingency tables.
    $p_{\mathrm{v}}^{\mathrm{adj}}$ are corrected for multiple comparison (Holm–Bonferroni correction).

    \begin{center}
        \footnotesize
        \captionof{table}{Ontology enrichment with gene-specific mutation-selection model}
        \input{gene-MutSel-0.025-table.core.tex}
    \end{center}
    The genes detected under adaptation by mutation-selection codon models are enriched primarily with ontologies related to immune system processes (innate immune response, immune response, inflammatory response, defense response to virus, etc) and ontologies related to the external membrane (extracellular region, extracellular exosome, etc).

    \subsection{Site-specific $\bm{\omega}$-based model}
    \label{subsec:w-based-codon-method}
    For each gene, we computed the proportion of sites classified under an adaptive regime such that the lower bound for the posterior credibility interval of site-specific $\omega$ ($\alpha=0.05$) is above 1.
    Because this is a unilateral test ($\omega > 1$) the risk is $\alpha/2=0.05/2=0.025$ for each site.
    For each ontology, the proportion of sites under adaptation is compared between the set of genes sharing this given ontology and the rest of the genes with Mann-Whitney U test.
    $p_{\mathrm{v}}^{\mathrm{adj}}$ are corrected for multiple comparison (Holm–Bonferroni correction).

    \begin{center}
        \footnotesize
        \captionof{table}{Ontology enrichment with site-specific $\bm{\omega}$-based model}
        \input{site-Classical-0.025-table.core.tex}
    \end{center}
    The sites detected under adaptation by site-specific $\omega$-based codon models are enriched primarily with ontologies related to immune system processes (innate immune response, immune response, inflammatory response, defense response to virus, etc) and ontologies related to the external membrane (extracellular region, extracellular exosome, etc).

    \subsection{Site-specific mutation-selection model}
    \label{subsec:mutation-selection-codon-method}

    For each gene, we computed the proportion of sites classified under an adaptive regime such that the lower bound for the posterior credibility interval of site-specific $\omega$ ($\alpha=0.05$) is above the upper bound of the posterior credibility interval of site-specific $\omega_{0}$ ($\alpha=0.05$).
    Because this is a unilateral test ($\omega > \omega_{0}$) and the two posterior credibility interval are independent, the risk is $(\alpha/2)^2=0.025^2=6.25 \times 10^{-4}$ for each site.
    For each ontology, the proportion of sites under adaptation is compared between the set of genes sharing this given ontology and the rest of the genes with Mann-Whitney U test.
    $p_{\mathrm{v}}^{\mathrm{adj}}$ are corrected for multiple comparison (Holm–Bonferroni correction).

    \begin{center}
        \footnotesize
        \captionof{table}{Ontology enrichment with site-specific mutation-selection model}
        \input{site-MutSel-0.025-table.core.tex}
    \end{center}
    The sites detected under adaptation by mutation-selection codon models are enriched primarily with ontologies related to immune system processes (innate immune response, immune response, inflammatory response, defense response to virus, etc) and ontologies related to the external membrane (extracellular region, extracellular exosome, etc).
    The ontologies are similar to the sites detected by site-specific $\omega$-based model, although with stronger statistical support (lower $p_{\mathrm{v}}^{\mathrm{adj}}$).
    Moreover, sites detected under adaptation by mutation-selection codon models are also enriched with ontologies related to oxidoreductase activites, oxidation-reduction process and mitochondrion, which are not found by site-specific $\omega$-based model.

    \subsection{Exclusive to site-specific mutation-selection model}
    \label{subsec:exclusive-to-mutation-selection-codon-method}

    For each gene, we computed the proportion of sites classified under an adaptive regime such that the lower bound for the posterior credibility interval of site-specific $\omega$ ($\alpha=0.05$) is above the upper bound of the posterior credibility interval of site-specific $\omega_{0}$ ($\alpha=0.05$), while the mean site-specific $\omega$ is below $1$.
    Thus, these sites cannot be detected by site-specific mutation-selection model.
    Because this is a unilateral test ($\omega > \omega_{0}$) and the two posterior credibility interval are independent, the risk is $(\alpha/2)^2=0.025^2=6.25 \times 10^{-4}$ for each site.
    For each ontology, the proportion of sites under adaptation is compared between the set of genes sharing this given ontology and the rest of the genes with Mann-Whitney U test.
    $p_{\mathrm{v}}^{\mathrm{adj}}$ are corrected for multiple comparison (Holm–Bonferroni correction).

    \begin{center}
        \footnotesize
        \captionof{table}{Ontology enrichment with sites exclusive to mutation-selection model}
        \input{site-MutSelExclu-0.025-table.core.tex}
    \end{center}

    The sites detected under adaptation solely by mutation-selection codon models (with $\omega < 1$) are enriched primarily with ontologies related to oxidoreductase activites, oxidation-reduction and mitochondrion process as well as ontologies related to the membrane (transmembrane transport, integral component of membrane, transmembrane transporter activity, etc).

    \pagebreak


    \section{Graphical abstract for the pipeline.}
    \label{subsec:method-summary}

    \begin{center}
        \captionof{figure}{~}
        \includegraphics[width=\linewidth]{pipeline}
    \end{center}

    \pagebreak


    \section{Rate of adaptation enrichment while controlling for $\bm{\omega}$}
    \label{sec:controlling-for-omega}

    $\omega$ is controlled to be the same in the nearly-neutral replicate and the adaptive set of genes, such as to alleviate the fact that genes classified as adaptive have a higher $\omega$ than genes classified as nearly-neutral, which could bias our comparison since $\rateApop$ could simply be higher for genes with higher $\omega$.

    \begin{center}
        \captionof{figure}{~}
        \includegraphics[width=\linewidth]{polymorphism-method-control}
    \end{center}

    The random sampling is weighted to control for $\omega$ in the set of nearly-neutral genes/sites.
    First, a normal distribution is fitted to $\omega$ in both sets, and the probability density is called $f$ for the adaptive set and $g$ for the nearly-neutral set.
    Secondly, for each gene/site classified as nearly-neutral the weight is computed as the ratio $f(\omega)/g(\omega)$ for this specific gene/site.
    Sampling with this procedure produce a set of genes/sites classified as nearly-neutral with the same $\omega$ on average than the set of adaptive genes/sites.
    \newpage
    \input{control/results_genes.tex}

    In \textit{Bos}, \textit{Capras} and \textit{Ovis}, $\dnds$ computed for the set of genes under adaptation at the phylogenetic scale (red, $1 > \omega > \omega_{0}$) is not higher than for the set of genes under a nearly-neutral regime (green, $1 > \omega \simeq \omega_{0}$), showing that the sampling procedure controlling for $\omega$ at the phylogenetic scale is valid.
    However, even though $\dnds$ is not higher (genes are not faster overall), the rate of adaptation at the population-genetic scale ($\rateApop$) for the set of genes supposedly under adaptation is higher than for the set of nearly-neutral genes in all populations.
    Altogether, these genes have a higher rate of adaptation, while not evolving faster than there nearly-neutral counterpart, showing that the rate of adaptation computed at the phylogenetic scale is able to detect genes under pervasive adaptation.

    \pagebreak

    \input{control/results_sites.tex}

    In \textit{Bos}, \textit{Capras}, \textit{Ovis}, \textit{Chlorocebus} and \textit{Homo}, $\dnds$ computed for the set of sites under adaptation (red, $1 > \omega > \omega_{0}$) at the phylogenetic scale is not higher than for the set of sites under a nearly-neutral regime (green, $1 > \omega \simeq \omega_{0}$), showing that the sampling procedure controlling for $\omega$ at the phylogenetic scale is valid.
    However, even though $\dnds$ is not higher (sites are not faster overall), the rate of adaptation at the population-genetic scale ($\rateApop$) for the set of sites supposedly under adaptation is higher than for the set of nearly-neutral sites in \textit{Bos} and \textit{Ovis}.
    Altogether, these sites have a higher rate of adaptation, while not evolving faster than there nearly-neutral counterpart, showing that the rate of adaptation computed at the phylogenetic scale is able to detect sites under pervasive adaptation, even though these sites are with $\omega < 1$ and thus can not be detected by site-specific $\omega$-based codon models.

    \pagebreak


    \section{Rate of adaptation enrichment with $\bm{\alpha=0.005}$}
    \label{sec:threshold}
    For each protein-coding DNA alignment, the Monte-Carlo Markov-Chain (MCMC) is run during $2000$ points using the \href{https://github.com/bayesiancook/bayescode}{BayesCode} software, after a burn-in of $1000$ points.
    The mean of $\omega$ and $\omega_{0}$ are computed across the MCMC (after burn-in), as well as the $99.5$\% posterior credibility interval ($\alpha=0.005$) for each gene and site, which is more stringent than the $95$\% interval ($\alpha=0.05$) as shown in the main manuscript.
    Genes and sites classified under an adaptive regime (in red) are rejecting the nearly-neutral assumption such that a lower bound for the posterior credibility interval of $\omega$ is above the upper bound of the posterior credibility interval of $\omega_{0}$, meaning $\omega > \omega_{0}$.
    Genes and sites are classified under a nearly-neutral regime (in green) if the average $\omega$ is within the posterior credibility interval of the $\omega_{0}$, and respectively the average $\omega_{0}$ is also within the posterior credibility interval of  $\omega$, meaning $\omega = \omega_{0}$.
    Genes and sites that do not fall in any of these categories are considered unclassified.

    \subsection{Scatterplot with $\bm{\alpha=0.005}$}

    \begin{center}
        \captionof{figure}{~}
        \begin{minipage}{0.32\linewidth}
            \includegraphics[width=\linewidth, page=1]{scatterplot-gene-MutSel-0.0025}
        \end{minipage}
        \llap{\raisebox{1.2cm}{\scriptsize A\hspace{4.7cm}}}\hfill
        \begin{minipage}{0.32\linewidth}
            \includegraphics[width=\linewidth, page=1]{scatterplot-site-MutSel-0.0025}
        \end{minipage}
        \llap{\raisebox{1.2cm}{\scriptsize B\hspace{4.7cm}}}\hfill
        \begin{minipage}{0.32\linewidth}
            \includegraphics[width=\linewidth, page=1]{scatterplot-site-MutSelExclu-0.0025}
        \end{minipage}
        \llap{\raisebox{1.2cm}{\scriptsize C\hspace{4.7cm}}}\hfill
    \end{center}

    $\omega$ estimated by the site model against $\omega_{0}$ calculated by the mutation-selection model.
    Scatter plot of $14,509$ genes in panel A, with $99.5$\% posterior credibility interval ($\alpha=0.005$).
    Density plot of sites in panel B and C.
    Genes or sites are then classified whether they detected as adaptive ($\omega > \omega_{0}$ in red) or nearly-neutral ($\omega \simeq \omega_{0}$ in green).
    In panel C, the set of sites detected exclusively by mutation-selection codon models have a mean $\omega < 1 $.

    \newpage
    \input{no_control/results.tex}

    With a threshold of $\alpha=0.005$ more stringent than as shown in the main manuscript ($\alpha=0.05$), the rate of false positive is mechanically lower both at the site and gene level.
    However, the statistical power to test for enrichment of adaptation is not necessarily higher since we are working with fewer data.
    We thus obtain high values of $\rateApop$ for the set of genes and sites supposedly under adaptation at the phylogenetic scale, which also leads to higher $\Delta \rateApop$ when compared to the the set of genes and sites supposedly under a nearly-neutral regime.
    Since we have more variance due to less data (violin plots are more extended for the set of nearly-neutral replicates), this higher value of the statistic $\Delta \rateApop$
    does not translate to lower $p_{\mathrm{v}}^{\mathrm{adj}}$.
    We thus have to find a compromise between having large enough dataset to perform reliable computation of $\rateApop$ and having a very low rate of false positive for sites and genes under adaptation at the phylogenetic scale.
    Hence in the manuscript we settled with a rate of $\simeq 1 \%$ FDR at the gene level and $\simeq 5 \%$ FDR at the site level.

    \pagebreak
    \section{Rate of adaptation enrichment with polyDFE}
    \label{sec:polyDFE}

    The probability to sample allele at a given frequency (before fixation or extinction) is informative of its scaled selection coefficient at the population scale ($\Spop$).
    Pooled across many sites, the SFS is thus informative on the underlying $\Spop$ of mutations, given we have a neutral expectation.
    In this configuration, a single $\Spop$ for all sampled mutations is biologically not realistic.
    Accordingly, a distribution of fitness effects of mutations (DFE) is assumed, usually modeled as a continuous distribution\cite{eyre-walker_distribution_2006, eyre-walker_estimating_2009}.
    In this study, we used the software polyDFE\cite{tataru_inference_2017, tataru_polydfe_2020} with model C (continous distribution) and D (discrete distribution).

    PolyDFE requires one SFS for non-synonymous mutations and one for synonymous mutations (neutral expectation), as well as the total number of sites on which each SFS has been sampled.
    From the SFS and the number of sites for both synonymous and non-synonymous changes, polyDFE estimates parameters of the DFE ($\phi (\Spop)$) using maximum likelihood.
    The estimated DFE allows to subsequently computed the rate of adaptive evolution $\rateApop^{\mathrm{polyDFE}}$.

    \subsubsection*{polyDFE Model C}
    In model C, the DFE ($\phi$) is given by a mixture of a $\Gamma$ and Exponential distributions, parameterized by $\Spop_d$ , $b$, $p_b$
    and $\Spop_b$ as:
    \begin{equation}
        \phi (\Spop) =
        \begin{dcases}
            \left( 1 - p_b \right) f_{\Gamma}(-\Spop; -\Spop_d, b) & \text{ if $\Spop \leq 0$,} \\
            p_b f_{e}(\Spop; \Spop_b) & \text{ if $\Spop > 0$,} \\
        \end{dcases}
    \end{equation}
    where $\Spop_d < 0 $ is the estimated mean of the DFE for $\Spop \leq 0$,
    $b > 0$ is the shape of the $\Gamma$ distribution fixed to $1.0$,
    $0 \leq p_b \leq 1$ is the estimated probability that $\Spop > 0$,
    $\Spop_b > 0$ is the mean of the DFE for $\Spop > 0$ fixed to $4.0$,
    and $f_{\Gamma}(x; m, b)$ is the density of the $\Gamma$ distribution with mean m and shape b, while $f_{e}(x; m)$ is the density of the Exponential distribution with mean $m$.

    \subsubsection*{polyDFE Model D}
    In model D, the DFE ($\phi$) is given as a discrete distribution, where the selection coefficients can take one of $\Spop_i$ distinct
    values, $1 \leq i \leq K$, where each value $\Spop_i$ has probability $p_i$, with
    \begin{equation}
        \sum\limits_{i=0}^{K} p_i = 1.
    \end{equation}

    \subsection{PolyDFE model C - including divergence data}
    The rate of adaptation $\rateApop^{\mathrm{polyDFE}}$ is computed as the difference between the total rate of evolution $\omega=\dnds$ obtained from divergence data and the rate of non-adaptive evolution ($\omega_{\mathrm{NA}}(\phi)$) obtained from polymorphism data as in \textcite{tataru_inference_2017}:
    \begin{align}
        \rateApop^{\mathrm{polyDFE}} &= \omega - \omega_{\mathrm{NA}}(\phi), \\
        &= \dnds - \omega_{\mathrm{NA}}(\phi).
    \end{align}
    Formally, $\omega_{\mathrm{NA}}(\phi)$ is computed as the average fixation probability of mutations ($\pfix (\Spop)$) over the probability distribution given by the DFE $\phi (\Spop)$, taken only for the negatively selected mutations ($\Spop < 0$) as:
    \begin{align}
        \omega_{\mathrm{NA}}(\phi) &= \int_{-\infty}^{0} \pfix (\Spop) \phi (\Spop) \der \Spop, \\
        &= \int_{-\infty}^{0} \frac{\Spop}{1 - \e^{-\Spop}} \phi (\Spop) \der \Spop, \\
        &= \int_{-\infty}^{0} \frac{\Spop}{1 - \e^{-\Spop}} \left( 1 - p_b \right) f_{\Gamma}(-\Spop; -\Spop_d, b) \der \Spop. \\
    \end{align}
    Altogether, $\rateApop^{\mathrm{polyDFE}}$ is given as:
    \begin{align}
        \rateApop^{\mathrm{polyDFE}} &= \dnds - \omega_{\mathrm{NA}}(\phi), \\
        &= \dnds - \int_{-\infty}^{0} \frac{\Spop}{1 - \e^{-\Spop}} \phi (\Spop) \der \Spop.
    \end{align}
    \input{polyDFE/modelC-results-div.tex}

    \begin{itemize}
        \item At the gene level, sets of genes supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.19,0.29]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.32,0.42]$ across the populations.
        \item At the site level, sets of sites supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.25,0.42]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.11,0.99]$ across the populations.
    \end{itemize}
    $\rateApop^{\mathrm{polyDFE}}$ computed with polyDFE is higher than $\rateApop$ computed as the \textcite{mcdonald_adaptative_1991} statistic (figure 3 and table 1 in the main manuscript), suggesting that polyDFE has higher sensitivity to detect adaptation at the population-genetic scale.
    However, the genes and sites supposedly under a nearly-neutral regime have all values of $\rateApop^{\mathrm{polyDFE}}$ greater than 0, suggesting that the higher sensitivity(true positive rate) also results in a lower specificity (true negative rate).
    Altogether, the statistical test for the enrichment of $\rateApop$ between the set of adaptive and nearly-neutral genes and sites gives similar results whether computed by polyDFE ($\rateApop^{\mathrm{polyDFE}}$) or as \textcite{mcdonald_adaptative_1991} statistic.

    \pagebreak

    \subsection{PolyDFE model C - polymorphism data alone with $\Spop > 0$}
    As in \textcite{tataru_inference_2017}, the rate of adaptation $\rateApop^{\mathrm{polyDFE}}$ can also be estimated from the polymorphism data alone, computed as the average fixation probability $\pfix (\Spop)$ over the probability given the DFE ($\phi (\Spop)$), only for the positively selected mutations ($\Spop > 0$), as:
    \begin{align}
        \rateApop^{\mathrm{polyDFE}} &= \int_{0}^{+\infty} \pfix (\Spop) \phi (\Spop) \der \Spop, \\
        &= \int_{0}^{+\infty} \frac{\Spop}{1 - \e^{-\Spop}} \phi (\Spop) \der \Spop, \\
        &= \int_{0}^{+\infty} \frac{\Spop}{1 - \e^{-\Spop}} p_b f_{e}(\Spop; \Spop_b) \der \Spop.
    \end{align}
    \input{polyDFE/modelC-results-0.tex}

    \begin{itemize}
        \item At the gene level, sets of genes supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.20,0.31]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.32,0.43]$ across the populations.
        \item At the site level, sets of sites supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.24,0.44]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.16,1.04]$ across the populations.
    \end{itemize}

    The estimation of $\rateApop^{\mathrm{polyDFE}}$ computed with polyDFE using polymorphism data alone is quite consistent with the estimation combining polymorphism and divergence (previous section).
    However, we higher variance in the estimation (violin plots are more extended for the set of nearly-neutral replicates, particularly noticeable for sites), reducing the statistical power for the enrichment test of $\rateApop^{\mathrm{polyDFE}}$ between the set of adaptive and nearly-neutral genes and sites.

    \pagebreak

    \subsection{PolyDFE model C - polymorphism data alone with $\Spop > 5$}

    The definition for a positively selected mutation by the criterion $\Spop > 0$ is also open to interpretation, and the integration limit can be set to a strictly positive value (e.g. 1, 3 or 5) instead of 0\cite{galtier_adaptive_2016, tataru_inference_2017}.
    The reasoning is that mutations with a positive selection coefficient that is not very large are not necessarily advantageous mutations.
    In \textcite{galtier_adaptive_2016}, the threshold $\Spop > 5$ is used and the rate of adaptation is thus:
    \begin{equation}
        \rateApop^{\mathrm{polyDFE}} = \int_{5}^{+\infty} \frac{\Spop}{1 - \e^{-\Spop}} p_b f_{e}(\Spop; \Spop_b) \der \Spop.
    \end{equation}
    \input{polyDFE/modelC-results-5.tex}

    \begin{itemize}
        \item At the gene level, sets of genes supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.12,0.19]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.20,0.26]$ across the populations.
        \item At the site level, sets of sites supposedly under a nearly-neutral regime have an average value of $\rateApop^{\mathrm{polyDFE}}$ (average across the replicates) in the range $[0.14,0.26]$ across the populations, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.09,0.63]$ across the populations.
    \end{itemize}

    The estimation of $\rateApop^{\mathrm{polyDFE}}$ computed with polyDFE using polymorphism data alone and with a bound of $\Spop > 5$ to consider a mutation as adaptive\cite{galtier_adaptive_2016} is lower than $\rateApop^{\mathrm{polyDFE}}$ computed with a bound of $\Spop > 0$\cite{tataru_inference_2017}.
    However, the genes and sites supposedly under a nearly-neutral regime still have values of $\rateApop^{\mathrm{polyDFE}}$ greater than 0, suggesting that  $\rateApop^{\mathrm{polyDFE}}$ still has a lower specificity than the \textcite{mcdonald_adaptative_1991} statistic.

    \pagebreak

    \subsection{PolyDFE model D - including divergence}
    Again, the rate of adaptation $\rateApop^{\mathrm{polyDFE}}$ is computed as the difference between the total rate of evolution $\omega=\dnds$ obtained from divergence data and the rate of non-adaptive evolution ($\omega_{\mathrm{NA}}(\phi)$) obtained from polymorphism data.
    Also, $\omega_{\mathrm{NA}}(\phi)$ is computed as the average fixation probability of mutations ($\pfix (\Spop)$) over the probability distribution given by the DFE $\phi (\Spop)$, taken only for the negatively selected mutations ($\Spop < 0$) as:
    \begin{align}
        \omega_{\mathrm{NA}}(\phi) &= \sum_{i=1}^{K} \mathbbm{1}_{]-\infty, 0]}(\Spop_i) \pfix (\Spop_i) p_i,  \\
        &= \sum_{i=1}^{K} \mathbbm{1}_{]-\infty, 0]}(\Spop_i) \frac{\Spop_i}{1 - \e^{-\Spop_i}} p_i,
    \end{align}
    where $\mathbbm{1}_{]-\infty, 0]}(\Spop)$ is the indicator function:
    \begin{equation}
        \mathbbm{1}_{]-\infty, 0]}(\Spop) =
        \begin{dcases}
            1 & \text{ if $\Spop \leq 0$,} \\
            0 & \text{ if $\Spop > 0$.} \\
        \end{dcases}
    \end{equation}
    Altogether, $\rateApop^{\mathrm{polyDFE}}$ is given as:
    \begin{align}
        \rateApop^{\mathrm{polyDFE}} &= \dnds - \sum_{i=1}^{K} \mathbbm{1}_{]-\infty, 0]}(\Spop_i) \frac{\Spop_i}{1 - \e^{-\Spop_i}} p_i.
    \end{align}
    \newpage
    \input{polyDFE/modelD-results-div.tex}

    \begin{itemize}
        \item At the gene level, $\rateApop^{\mathrm{polyDFE}}$ for nearly-neutral genes is the range $[-0.01,0.27]$, while $\rateApop^{\mathrm{polyDFE}}$ for genes under adaptation (at the phylogenetic scale) is in the range $[0.04,42]$.
        \item At the site level, $\rateApop^{\mathrm{polyDFE}}$ for nearly-neutral sites is the range $[0.0,0.34]$, while $\rateApop^{\mathrm{polyDFE}}$ for sites under adaptation (at the phylogenetic scale) is in the range $[0.40,0.98]$.
    \end{itemize}
    The estimation of $\rateApop^{\mathrm{polyDFE}}$ computed with polyDFE using polymorphism and divergence data with either a the model C (continous DFE) or D (discrete DFE) are quite different in absolute value.
    The underlying assumption for the mathematical constrains on the DFE thus have a large impact on the estimation of $\rateApop^{\mathrm{polyDFE}}$ while the true underlying
    DFE is unknown.

    \pagebreak

    \printbibliography

\end{document}